// yylib-react/dist/business/bpmapprove/YYBatchApprove.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

require('./bpm/styles/less/bpmtask.less');

var _antd = require('antd');

var _reactCustomScrollbars = require('react-custom-scrollbars');

var _BPMURL = require('./bpm/url/BPMURL');

var _BPMURL2 = _interopRequireDefault(_BPMURL);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _require = require('yylib-ui'),
    YYReferInput = _require.YYReferInput,
    YYButton = _require.YYButton,
    YYModal = _require.YYModal,
    YYTableCol = _require.YYTableCol,
    YYTable = _require.YYTable;

var ajax = require('yylib-utils/ajax');
var AuthToken = require("yylib-utils/AuthToken");

var FormItem = _antd.Form.Item;
var OptGroup = _antd.Select.OptGroup;
var RadioGroup = _antd.Radio.Group;
var _this = null;
var mapStateToProps = function mapStateToProps(state) {

		return {
				bpm: state.bpm
		};
};
var columns = [{
		title: '单据类型',
		dataIndex: 'billtypename'

}, {
		title: '单据编号',
		dataIndex: 'processInsName'
}, {
		title: '原因',
		dataIndex: 'msg'
}];

var YYBatchApprove = function (_React$Component) {
		_inherits(YYBatchApprove, _React$Component);

		function YYBatchApprove(props) {
				_classCallCheck(this, YYBatchApprove);

				var _this2 = _possibleConstructorReturn(this, (YYBatchApprove.__proto__ || Object.getPrototypeOf(YYBatchApprove)).call(this, props));

				_this = _this2;
				_this2.doApprove = _this2.doApprove.bind(_this2);
				_this2.cancel = _this2.cancel.bind(_this2);
				_this2.beforeApprove = _this2.beforeApprove.bind(_this2);
				_this2.state = {
						opinion: '同意',
						customTip: '',
						showApprove: false,
						placeholder: "请输入审批语",
						errRecord: [],
						loadflag: false
				};
				return _this2;
		}

		_createClass(YYBatchApprove, [{
				key: 'inputOnChange',
				value: function inputOnChange(e) {
						this.setState({
								customTip: e.target.value
						});
				}
				//初始化渲染执行之后立刻调用一次

		}, {
				key: 'componentDidMount',
				value: function componentDidMount() {
						AuthToken.init();
				}

				//数据更新后通知此方法

		}, {
				key: 'componentWillReceiveProps',
				value: function componentWillReceiveProps(nextProps) {}

				//显示用户参照

		}, {
				key: 'beforeApprove',
				value: function beforeApprove(e) {}

				//审批操作

		}, {
				key: 'doApprove',
				value: function doApprove() {
						var userId = this.state.userId; //登  
						var opinionContent = "同意";
						var comment = this.state.customTip ? this.state.customTip : "";
						if (comment == "") {
								comment = opinionContent;
						}
						var approveObject = [];
						approveObject = {
								userId: AuthToken.getUserId(),
								bills: this.props.bills,
								approveType: opinionContent,
								comment: comment
						}; //回调数据
						this.onBpmApprove(approveObject);
				}
				//提交审批

		}, {
				key: 'onBpmApprove',
				value: function onBpmApprove(data) {
						var _this = this;
						var dispatch = this.props.dispatch;
						var serUrl = this.props.serUrl;

						var flag = false; //用来标识是否调用服务成功
						var errRecord = [];
						var url = serUrl + _BPMURL2.default.BPM_BATCHAPPROVE;
						console.log(data, "data");
						ajax.postJSON(url, data, null, null, function (result) {
								if (result) {
										var text = eval("(" + result.text + ")");
										console.log(text, "-------------------");
										_this.props.dataSource = [];

										if (Array.isArray(text) && text.length > 0) {
												flag = text[0].success;
										}
										if (Array.isArray(text) && text.length > 0) {
												text.map(function (value) {
														var record = {};
														flag = flag && value.success;
														if (!value.success) {
																record.billtypename = value.bill.billtypename;
																record.processInsName = value.bill.processInsName;
																record.msg = value.msg;
																errRecord.push(record);
														}
												});
												_this.setState({
														errRecord: errRecord
												});
										}
										if (flag && _this.state.errRecord.length == 0) {
												_this.cancel();
												_antd.message.info("批审成功！");
										}
								} else {
										_this.cancel();
										_antd.message.info("网络可能有问题，批审失败！");
								}
						});

						this.setState({ loadflag: true });
				}

				//取消审批

		}, {
				key: 'cancel',
				value: function cancel(e) {

						if (this.props.closeApprove) {
								this.props.closeApprove();
						}
						this.state.errRecord = [];
						this.state.loadflag = false;
						this.state.customTip = "";
				}
		}, {
				key: 'render',
				value: function render() {
						var content = "";
						var radioText = null;
						var formItemLayout = {
								labelCol: { span: 4 },
								wrapperCol: { span: 19 }
						};
						var customTip = this.state.customTip;

						var approveTitle = [_react2.default.createElement(
								'div',
								null,
								'\u6279\u91CF\u5BA1\u6279'
						)];
						var recordTitle = [_react2.default.createElement(
								'div',
								null,
								'\u6279\u91CF\u5BA1\u6279\u5931\u8D25\u8BB0\u5F55'
						)];
						var mainRecordModal = _react2.default.createElement(
								'div',
								{ style: { padding: 15 } },
								_react2.default.createElement(
										_reactCustomScrollbars.Scrollbars,
										{ style: { height: 300 } },
										_react2.default.createElement(YYTable, { bordered: true, columns: columns, dataSource: this.state.errRecord, pagination: false })
								)
						);
						var footRecordModal = [_react2.default.createElement(
								YYButton,
								{ type: 'primary', size: 'large', onClick: this.cancel },
								'\u6211\u77E5\u9053\u4E86'
						)];
						var mainApproveModal = _react2.default.createElement(
								'div',
								{ className: content },
								_react2.default.createElement(
										_reactCustomScrollbars.Scrollbars,
										{ style: { height: 300 } },
										_react2.default.createElement(
												_antd.Form,
												{ horizontal: true, form: this.props.form, className: 'yyui-bpm-bpmtask-form' },
												_react2.default.createElement(
														_antd.Row,
														null,
														_react2.default.createElement(
																_antd.Col,
																{ span: '23' },
																_react2.default.createElement(
																		FormItem,
																		_extends({ className: 'yyui-bpm-form-item',
																				label: '\u5BA1\u6279\u610F\u89C1\uFF1A'
																		}, formItemLayout),
																		_react2.default.createElement(
																				RadioGroup,
																				{ defaultValue: 1 },
																				_react2.default.createElement(
																						_antd.Radio,
																						{ value: 1 },
																						'\u540C\u610F'
																				)
																		)
																)
														)
												),
												_react2.default.createElement('hr', { className: 'yyui-bpm-cut-off' }),
												_react2.default.createElement(
														_antd.Row,
														null,
														_react2.default.createElement(
																_antd.Col,
																{ span: '23' },
																_react2.default.createElement(
																		FormItem,
																		_extends({ className: 'yyui-bpm-ant-form-item',
																				id: 'control-textarea',
																				label: '\u5BA1\u6279\u8BED\uFF1A'
																		}, formItemLayout),
																		_react2.default.createElement(_antd.Input, { type: 'textarea', placeholder: this.state.placeholder, value: customTip, onChange: this.inputOnChange.bind(this), className: 'yyui-bpm-textarea-input', rows: '5' })
																)
														)
												)
										)
								)
						);
						var footApproveModal = [_react2.default.createElement(
								'div',
								{ className: 'yyui-bpm-footBtn' },
								_react2.default.createElement(
										'div',
										{ className: 'yyui-bpm-operateBtn' },
										_react2.default.createElement(
												'div',
												{ className: 'yyui-bpm-cancel' },
												_react2.default.createElement(
														YYButton,
														{ onClick: this.cancel, 'data-id': 'cancel', type: 'default', disabled: this.state.loadflag },
														'\u53D6\u6D88'
												)
										),
										_react2.default.createElement(
												'div',
												{ className: 'yyui-bpm-comfirm' },
												_react2.default.createElement(
														YYButton,
														{ onClick: this.doApprove, type: 'primary', loading: this.state.loadflag, disabled: this.state.loadflag },
														'\u786E\u5B9A'
												)
										)
								)
						)];

						return _react2.default.createElement(
								'div',
								{ className: 'yyui-bpm-main' },
								_react2.default.createElement(
										_antd.Modal,
										{ style: { top: 50 },
												title: !this.state.errRecord.length ? approveTitle : recordTitle,
												width: 630,
												visible: this.props.showApprove,
												closable: false,
												className: 'yyui-bpm-main',
												footer: !this.state.errRecord.length ? footApproveModal : footRecordModal },
										!this.state.errRecord.length ? mainApproveModal : mainRecordModal
								)
						);
				}
		}]);

		return YYBatchApprove;
}(_react2.default.Component);

;
module.exports = (0, _reactRedux.connect)(mapStateToProps)(YYBatchApprove);
