// yylib-react/dist/yylib2/formcontrols/tag/YYTagItem.js
'use strict';

var React = require('react');
var YYClass = require('../../base/YYClass');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
var classnames = require('classnames');
var _ = require('lodash');
require('./tagitem.less');
//定义META
var uiMeta = {
    name: 'tagitem',
    fields: [
    // field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('text', null, PropTypes.string, true, '显示文字'), field('value', null, PropTypes.string, true, '用于表单取值,唯一'), field('checked', undefined, PropTypes.string, true, '是否默认选中'), field('disabled', false, PropTypes.string, true, '是否启用禁用状态')]
};

var YYTagItem = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        return {
            checked: this.props.checked
        };
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('checked' in nextProps) {
            this.setState({
                checked: nextProps.checked
            });
        }
    },
    clickItem: function clickItem(value, text, disabled) {
        if (disabled == true) return;
        this.setState({
            checked: !this.state.checked
        });
        if (_.isFunction(this.props.onChange)) {
            this.props.onChange(value, text, !this.state.checked);
        }
    },
    render: function render() {
        var _props = this.props,
            value = _props.value,
            text = _props.text,
            disabled = _props.disabled;

        var cls = classnames({
            'yyui-tagitem': true,
            'tagitem-active': this.state.checked,
            'tagitem-disabled': disabled
        });
        return React.createElement(
            'span',
            { className: cls,
                onClick: this.clickItem.bind(null, value, text, disabled),
                value: value },
            text
        );
    }
});

module.exports = YYTagItem;
