// yylib-cscec/dist/template/resource/BaseHost.js
"use strict";

/**
 * Created by xg on 2017/3/23.
 *
 * 依据环境自判断IP地址
 */
var ADDR; // 环境地址
var SHARE_BASEHOST; // share工程基准地址
var SUPPORT_BASEHOST; // support工程基准地址
var BPM_SERVER_URL; // BPM服务地址;
var ATTACH_SERVER_URL; // 附件管理服务地址
var URL_HOME_PORTAL; // Portal首页地址
var SUPPORT_WEB_URL; // support模块服务地址
var SHARE_WEB_URL; // share模块服务地址
var PRINT_PREVIEW_URL; // 打印预览服务地址

window.EnvConfig = window.EnvConfig || null;
console.log("[yylib-cscec]_EnvConfig", window.EnvConfig);
if (window.EnvConfig && window.EnvConfig.YYLIB_BASEHOST) {
    //配置外部化
    ADDR = window.EnvConfig.YYLIB_BASEHOST;
    SHARE_BASEHOST = window.EnvConfig.SHARE_BASEHOST;
    SUPPORT_BASEHOST = window.EnvConfig.SUPPORT_BASEHOST;
} else if (process.env.YYLIB_ENV == 'prod') {
    // 生产环境
    ADDR = "http://zjy.cscec.com.cn";
} else if (process.env.YYLIB_ENV === 'outer' || process.env.YYLIB_ENV === 'test') {
    // 阿里云环境
    ADDR = "http://60.205.12.0:81";
} else if (process.env.YYLIB_ENV === 'dev') {
    // 开发环境
    ADDR = "https://dev.yonyouccs.com";
} else if (process.env.YYLIB_ENV === 'testcscec3b') {
    // 3局测试环境
    ADDR = "//test.cscec3b.com.cn";
} else {
    ADDR = "https://dev.yonyouccs.com";
}

if (!SHARE_BASEHOST) {
    SHARE_BASEHOST = ADDR;
}
if (!SUPPORT_BASEHOST) {
    SUPPORT_BASEHOST = ADDR;
}
BPM_SERVER_URL = ADDR;
ATTACH_SERVER_URL = ADDR;
URL_HOME_PORTAL = ADDR + "/icop-workbench/#home";
SUPPORT_WEB_URL = SUPPORT_BASEHOST + "/icop-support-web";
SHARE_WEB_URL = SHARE_BASEHOST + "/icop-share-web";
PRINT_PREVIEW_URL = ADDR + "/icop-print/print/getPreviewHtml";

module.exports = {
    ADDR: ADDR,
    SHARE_BASEHOST: SHARE_BASEHOST,
    SUPPORT_BASEHOST: SUPPORT_BASEHOST,
    BPM_SERVER_URL: BPM_SERVER_URL,
    ATTACH_SERVER_URL: ATTACH_SERVER_URL,
    URL_HOME_PORTAL: URL_HOME_PORTAL,
    SUPPORT_WEB_URL: SUPPORT_WEB_URL,
    SHARE_WEB_URL: SHARE_WEB_URL,
    PRINT_PREVIEW_URL: PRINT_PREVIEW_URL
};
