// yylib-cscec/dist/template/resource/PubConstants.js
"use strict";

/**
 * Created by xg on 2017/3/24.
 * 建造模板常量
 */
var PubConstants = {

    /**
     * 设计器中的组件
     */
    COMP_ACCORDION: "YYAccordion", // 多卡片组件
    COMP_NORMAL_TOOLBAR: "YYToolbar", // 普通工具栏
    COMP_FORM: 'FormWidget', // 普通表单
    COMP_FORM_ITEM: "FormItemWidget", // 表单项
    COMP_EDIT_TABLE: 'EditTableWidget', // 编辑表格
    COMP_EDIT_ITEM: "EditItemWidget", // 编辑表格列
    COMP_DATA_TABLE_COL: "YYTableCol", // 数据表格列
    COMP_TABS: "YYTabs", // 多页签
    COMP_PLUGIN: "CPPluginMgr", // 插件组件
    COMP_DATA_TABLE: 'CPListTable', // 数据表格
    COMP_DATA_TABLE2: 'TableWidget', // 数据表格
    COMP_BASIC_TABLE: 'YYTable', // 基础表格
    COMP_AUDIT_FORM: 'CPAuditInfo', // 审计信息表单
    COMP_BPM_SUBMIT: 'CPBpmSubmit', // 提交组件
    COMP_BPM_APPROVE: 'CPBpmApprove', // 审批组件
    COMP_ATTACH: 'CPAttachMgr', // 附件管理组件
    COMP_PRINT: 'CPPrintButton', // 打印组件
    COMP_CARD_TITLE: "CPCardTitle", // 卡片界面标题组件
    COMP_CARD_PAGE: "CPCardPage", // 卡片界面外层容器组件
    COMP_LIST_PAGE: "CPListPage", // 列表界面外层容器组件
    COMP_SEARCH_BAR: "CPSearchBar", // 列表界面的搜索栏
    COMP_SCHEME_GROUP: "YYSchemeGroup", // 列表界面的搜索方案栏
    COMP_NORMAL_BTN: "YYButton", // 编辑表格工具栏按钮
    COMP_LIST_TOOLBAR: "CPListToolbar", // 列表页面工具栏
    COMP_SPECIAL_TAG: "specialTag", // 组件特殊标识属性

    /**
     * 卡片
     */
    SAVE_BTN_KEY: "saveBtn", // 保存按钮键值
    SAVE_ADD_BTN_KEY: "saveAddBtn", // 保存并新增按钮键值
    SAVE_COPY_BTN_KEY: "saveCopyBtn", // 保存并复制按钮键值
    DELETE_BTN_KEY: "removeBtn", // 删除按钮键值
    RELOAD_BTN_KEY: "reloadBtn", // 刷新按钮键值
    COPY_BTN_KEY: "copyBtn", //  复制按钮键值
    PK_FIELD: "primaryKey", // 审计信息中的主键字段名称
    TS_FIELD: "ts", // 审计信息中的ts字段名称
    BILL_TYPE_FIELD: "billType", // 审计信息中的单据类型code字段名称
    BILL_TYPE_NAME_FIELD: "billName", // 审计信息中的单据类型name字段名称
    BILL_STATE_FIELD: "billState", // 审计信息中的单据状态字段名称
    BILL_CODE_FIELD: "billCode", // 审计信息中的单据编号字段名称
    ATTACH_MGR_FIELD: "attachMgr", // 审计信息中的附件管理字段(存储附件的id)

    /**
     * 列表
     */
    LIST_ATTACH_MGR: "listAttachMgr", // 附件管理自定义组件键值
    LIST_VIEW_APPROVE: "listViewApprove", // 查看审批流自定义组件键值
    LIST_EXPORT_EXCEL: "listExportExcel", // 导出Excel 自定义组件键值
    LIST_COLUMN_CONFIG: "columnConfig", // 表格列配置自定义组件键值
    LIST_COMMON_PLUGIN: "commonPlugin", // 列表插件键值
    LIST_BUTTON_EXPORT: "exportBtn", // 列表导出按钮键值
    LINE_BTN_IDENTIFY: "lineBtn", // 行操作按钮标识
    LINE_BTN_EDIT: "editLineBtn", // 行编辑
    LINE_BTN_DEL: "delLineBtn", // 行删除
    LINE_BTN_ATTACH: "lineAttachBtn", // 行附件
    LINE_VIEW_APPROVE: "lineApproveBtn", // 查看审批流
    LIST_ITEM_VISIBLE_PROP: "isShow", // 数据表格的字段显示属性名
    LIST_OPERATE_KEY: "operate", // 数据表格操作列的键值

    /**
     * 编辑表格
     */
    DATE_ITEM_TYPE: "date", // 日期框ui类型
    ATTACH_ITEM_TYPE: "attachupload", // 上传附件ui类型
    ATTACH_MGR_ITEM_TYPE: "attachmgrw", // 附件管理UI类型
    IMAGE_ITEM_TYPE: "imageupload", // 上传图片ui类型
    ATTACH_UPLOAD_FIELD: "attachObjs", // 附件上传与后端对应的字段

    /**
     * 公共
     */
    DEFAULT_PK_FIELD_VALUE: "id", // 默认的主键字段值
    DEFAULT_DATE_FORMAT: "yyyy-MM-dd HH:mm:ss", // 默认的日期格式
    MSG_TIME_OUT: 3, // 消息提示延时时间(秒)
    LOADING_TIME_OUT: 1000, // 卡片界面加载中延时时间(毫秒)
    DEFAULT_ORG_FIELD: "companyId", // 单据上面默认的组织字段
    DEFAULT_FILE_SOURCE_TYPE: "102345678", // 默认的附件sourceType

    /**
     * 列表插件属性
     */
    PLUGIN_ATTACH_MGR: "attachMgr", // 附件管理
    PLUGIN_VIEW_APPROVE: "viewApprove", // 查看审批历史
    PLUGIN_EXPORT_EXCEL: "exportExcel", // 导出excel
    PLUGIN_COLUMN_CONFIG: "columnConfig", // 表格列配置

    /**
     * 单据状态
     */
    BILL_STATE: {
        FREE: 0, // 自由态
        COMMIT: 1, // 提交态
        APPROVING: 2, // 审批进行中
        APPROVED: 3, // 审批通过
        UNAPPROVED: 4, // 审批不通过
        APPROVED2: 5 // 已审批(直审)
    },

    /**
     * 注册按钮包含的属性
     */
    REGISTER_BTN_PROP: {
        BTN_CODE: 'btnCode', // 按钮编码
        BTN_NAME: 'btnName', // 按钮名称
        ENABLE: 'enable' // 是否启用
    },

    /**
     * 精度相关
     */
    NUMBER_ITEM_TYPE: "number", // 整数框ui类型
    PRECISION_ITEM_TYPE: "currency", // 数值框ui类型
    PRECISION_CONFIG: {
        NONE: { // 无
            TYPE: "none",
            PROPERTY: "none",
            VALUE: "00"
        },
        WEIGHT: { // 重量
            TYPE: "weight",
            PROPERTY: "weight_precision",
            VALUE: "01"
        },
        NUM: { // 数量
            TYPE: "num",
            PROPERTY: "num_precision",
            VALUE: "02"
        },
        MONEY: { // 金额
            TYPE: "money",
            PROPERTY: "money_precision",
            VALUE: "03"
        }
    },

    /**
     * 导出
     */
    EXPORT: {
        // 数据范围
        SCOPE: {
            LINE: "line", // 选中行
            PAGE: "page", // 当前页面
            ALL: "all" // 所有
        },
        // 默认后缀
        SUFFIX: ".xlsx",
        // 默认日期格式
        DATE_FORMAT: "yyyy-MM-dd",
        // 业务异常code
        BUSINESS_EXCEPTION_CODE: 400,
        // 业务异常提示消息
        BUSINESS_EXCEPTION_MSG: "获取数据源出现异常"
    },

    /**
     * 页面类型
     */
    UI_TYPE: {
        LIST: "list", // 列表
        CARD: "card" // 卡片
    },

    /**
     * 参照
     */
    REFER: {
        ID_FIELD: "id",
        CODE_FIELD: "code",
        NAME_FIELD: "name"
    },

    PRECISION_SUFFIX: "_precision_config", // 浏览器本地缓存中的精度参数后缀

    ENABLE_BTN_PERMISSION: false //是否启用按钮权限(弃用)
};
module.exports = PubConstants;
