// yylib-react/dist/business/attach/mgr/AttachListModal.js
'use strict';

/**
 * Created by Administrator on 2016/7/26.
 */
var React = require('react');
var AttachTable = require('./AttachListTable');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass,
    YYModal = _require.YYModal,
    YYButton = _require.YYButton,
    YYTree = _require.YYTree,
    YYTreeNode = _require.YYTreeNode,
    YYCol = _require.YYCol,
    YYRow = _require.YYRow;

var AttachmentListModal = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        return { fileTypes: '' };
    },
    getInitialState: function getInitialState() {
        return {
            serverUrl: '', //接口
            fileUrl: "", //文件
            visible: false,
            sourceId: null,
            billType: null,
            sourceType: null,
            loading: false,
            value: '',
            defaultKey: ''
        };
    },

    componentWillMount: function componentWillMount() {
        var fileTypeMap = this.props.fileTypes;
        if (fileTypeMap) {
            var defaultKey = fileTypeMap[0].code;
            this.setState({ defaultKey: defaultKey, fileTypes: fileTypeMap, sourceType: defaultKey });
        } else {
            this.setState({ sourceType: '102345678' });
        }
    },
    handleCancel: function handleCancel() {
        this.props.onCancel();
    },
    onSelectNode: function onSelectNode(selectedKeys, info) {
        //console.log(selectedKeys,info);
        var key = selectedKeys[0];
        this.setState({ sourceType: key });
    },
    render: function render() {
        var that = this;
        var nameMap = this.props.fileTypes;
        //是否定义文件分类
        var showTree = false;
        var tableWidth = 24;
        if (nameMap) {
            showTree = true;
            tableWidth = 19;
        }
        var defaultKey = this.state.defaultKey;
        var loop = function loop(data) {
            return data.map(function (item) {
                return React.createElement(YYTreeNode, { key: item.code, title: item.name });
            });
        };
        return React.createElement(
            YYModal,
            { ref: 'modal',
                wrapClassName: 'vertical-center-infomodal',
                visible: this.props.visible,
                title: '\u9644\u4EF6\u7BA1\u7406', width: '800',
                onCancel: this.handleCancel, footer: [React.createElement(
                    YYButton,
                    { key: 'back', onClick: this.handleCancel },
                    '\u5173\u95ED'
                )] },
            React.createElement(
                YYRow,
                { className: 'batch-opt' },
                showTree ? React.createElement(
                    YYCol,
                    { span: '5' },
                    React.createElement(
                        YYTree,
                        { defaultSelectedKeys: defaultKey, onSelect: this.onSelectNode },
                        loop(nameMap)
                    )
                ) : '',
                React.createElement(
                    YYCol,
                    { span: tableWidth },
                    React.createElement(AttachTable, { serverUrl: this.props.serverUrl, fileUrl: this.props.fileUrl, readOnly: true, modal: this, billType: this.props.billType, showUploadBtn: this.props.showUploadBtn, showDelBtn: this.props.showDelBtn, sourceId: this.props.sourceId, sourceType: this.state.sourceType, pageVisible: true })
                )
            )
        );
    }
});
module.exports = AttachmentListModal;
