// yylib-react/dist/business/attach/mgr/AttachMgrTable.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * Created by Administrator on 2016/7/26.
 */
var React = require('react');

var _require = require('yylib-ui'),
    YYCol = _require.YYCol,
    YYRow = _require.YYRow,
    YYButton = _require.YYButton,
    YYTable = _require.YYTable,
    YYUpload = _require.YYUpload,
    YYClass = _require.YYClass,
    YYNotice = _require.YYNotice,
    YYToolbar = _require.YYToolbar,
    YYMessage = _require.YYMessage;

require("../upload.css");

var _require2 = require("../attachaction"),
    loadAttachList = _require2.loadAttachList,
    delAttach = _require2.delAttach,
    showErrMsg = _require2.showErrMsg,
    downloadUrl = _require2.downloadUrl,
    listUrl = _require2.listUrl,
    uploadUrl = _require2.uploadUrl,
    delUrl = _require2.delUrl,
    handleUrl = _require2.handleUrl,
    formatDate = _require2.formatDate;

var pageSize = 10; //每页显示条数
var AuthToken = require("yylib-utils/AuthToken");

var _require3 = require('yylib-utils/MathUtil'),
    execute = _require3.execute;

var _ = require('lodash');
//下载url
//const downloadUrl = "/icop-file/file/download?id=";
//获取附件列表的url
//const listUrl = "/icop-file/file/list";
//上传url
//const uploadUrl = "/icop-file/file/upload2";

//const delUrl = "/icop-file/file/del";

//参数object&操作方法
//var dataMap = {};

var AttachMgrTable = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        return { showUploadBtn: true, showDelBtn: true, readOnly: false, beforeUpload: null, beforeDel: null,
            userId: '', userName: '', enableUser: true };
    },
    getInitialState: function getInitialState() {
        var enableUser = false;
        if (this.props.enableUser != false) {
            enableUser = true;
        }
        var userId = this.props.userId;
        var userName = this.props.userName;
        if (enableUser) {
            if (!userId) {
                userId = AuthToken.getUserId();
            }
            if (!userName) {
                userName = AuthToken.getUserName();
            }
        }
        return {
            serverUrl: '',
            selectId: [],
            current: 1,
            sourceId: null,
            billType: null,
            sourceType: null,
            pageVisible: false,
            totalsize: 0,
            data: [],
            pageInfo: {
                pageNumber: 1,
                pageSize: 10
            },
            isload: false,
            dataMap: {},
            selectItems: [],
            //上传进度
            //uploadEvent:null,
            fileList: [],
            userId: userId,
            userName: userName
        };
    },

    _loadList: function _loadList(serverUrl, billType, sourceId, sourceType) {
        var url = serverUrl + listUrl;
        var that = this;
        loadAttachList(url, { id: sourceId, billType: billType, type: sourceType }, function (data) {
            that.setState({ data: data });
            that.state.dataMap[sourceType] = data;
        }, function (msg) {
            showErrMsg("上传失败", msg);
        });
    },
    _removeFile: function _removeFile(serverUrl, sourceId, billType, sourceType, ids) {
        var url = serverUrl + delUrl;
        var that = this;
        if (sourceId && billType) {
            delAttach(url, { id: sourceId, billType: billType, sourceType: sourceType, attachIds: ids }, function (data) {
                if (data.success) {
                    that.doRemove(sourceType, ids);
                } else {
                    showErrMsg("删除失败", data.backMsg);
                }
            }, function (msg) {
                showErrMsg("删除失败", msg);
            });
        } else {
            that.doRemove(sourceType, ids);
        }
    },
    doRemove: function doRemove(sourceType, idArr) {
        var newFileArr = [];
        var files = this.state.data;
        var that = this;
        files.map(function (item, i) {
            if (!that.arrayContains(idArr, item.gid)) {
                newFileArr.push(item);
            }
        });
        this.setState({ data: newFileArr, selectId: [], selectItems: [] });
        this.state.dataMap[sourceType] = newFileArr; //将更新的数据放入map中
        this.onChangePage(1);

        this.refreshValue();
    },
    arrayContains: function arrayContains(arr, obj) {
        var i = arr.length;
        while (i--) {
            if (arr[i] === obj) {
                return true;
            }
        }
        return false;
    },
    //查询数据
    componentDidMount: function componentDidMount() {
        //进入编辑页面第一次时需要调用该方法加载
        var sourceId = this.props.sourceId;
        var billType = this.props.billType;
        var sourceType = this.props.sourceType;
        this.setState({ sourceId: sourceId });

        if (sourceId && billType) {
            this.setState({ data: [], dataMap: {} });
            this._loadList(this.props.serverUrl, billType, sourceId, sourceType);
        }
        /*var sourceId =this.props.sourceId;
        var billType=this.props.billType;
        var sourceType = this.props.sourceType;
        //this.setState({dataMap:[]});
        var lastSourceId = this.state.sourceId;
        if(!sourceId){
            sourceId="";
        }
        if(!lastSourceId){
            lastSourceId = "";
        }
        this.setState({sourceId:sourceId});
        if(sourceId && billType){
            this.setState({isload:true});
            if(sourceId!=lastSourceId){////sourceId变化时才加载附件列表
                this.setState({data:[],dataMap:{}});//sourceId有变化，该值已无意义
                this._loadList(this.props.serverUrl,billType,sourceId,sourceType);
            }
        }else{
            if(sourceId==lastSourceId){ //sourceId发生变化就应该清空列表
                var data = [];
                if(sourceType in this.state.dataMap) {
                    data = this.state.dataMap[sourceType];
                }
                this.setState({data: data});
            }else{
                this.setState({data:[],dataMap:{}});
            }
        }*/
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        /* if(this.state.isload){
             return;
         }*/
        var sourceId = nextProps.sourceId;
        var billType = nextProps.billType;
        var sourceType = nextProps.sourceType;
        //this.setState({dataMap:[]});
        var lastSourceId = this.state.sourceId;
        var lastSourceType = this.state.sourceType;

        if (!sourceId) {
            sourceId = "";
        }
        if (!lastSourceId) {
            lastSourceId = "";
        }
        if (!billType) {
            billType = "";
        }
        if (!lastSourceType) {
            lastSourceType = "";
        }
        this.setState({ sourceId: sourceId, sourceType: sourceType });
        if (sourceId && billType) {
            this.setState({ isload: true });
            if (sourceId != lastSourceId) {
                ////sourceId变化时才加载附件列表
                this.setState({ data: [], dataMap: {} }); //sourceId有变化，该值已无意义
                this._loadList(this.props.serverUrl, billType, sourceId, sourceType);
            } else {
                if (sourceType != lastSourceType) {
                    //文件类型变化了，重新加载
                    this.setState({ data: [] }); //sourceId有变化，该值已无意义
                    this._loadList(this.props.serverUrl, billType, sourceId, sourceType);
                }
            }
        } else {
            if (sourceId == lastSourceId) {
                //sourceId发生变化就应该清空列表
                if (sourceType != lastSourceType) {
                    //文件类型变化了，重新加载
                    var data = [];
                    if (sourceType in this.state.dataMap) {
                        data = this.state.dataMap[sourceType];
                    }
                    this.setState({ data: data });
                }
            } else {
                this.setState({ data: [], dataMap: {} });
            }
        }
    },

    //显示条数
    showTotal: function showTotal(total) {
        return '\u5171 ' + total + ' \u6761';
    },
    //改变页码
    onChangePage: function onChangePage(page) {
        var pageInfo = {
            pageNumber: page - 1,
            pageSize: this.state.pageInfo.pageSize
        };
        this.setState({
            pageInfo: pageInfo
        }, function () {});
    },
    //批量删除
    deleteAll: function deleteAll(e) {
        var ids = this.state.selectId;
        if (ids && ids.length > 0) {
            //删除前判断
            if (this.props.beforeDel) {
                if (!this.props.beforeDel(this.state.selectItems)) {
                    return false;
                }
            }
            this._removeFile(this.props.serverUrl, this.props.sourceId, this.props.billType, this.props.sourceType, ids);
        } else {
            YYMessage.warn("请选择要删除的文件");
        }
    },
    onUpload: function onUpload(e) {
        e.preventDefault();
    },
    refreshValue: function refreshValue() {
        // if(this.props.sourceId){
        //     this.props.modal.changeValue(null);
        //     return;
        // }
        var fileLength = 0;
        var map = {};
        for (var prop in this.state.dataMap) {
            if (this.state.dataMap.hasOwnProperty(prop)) {
                var arr = this.state.dataMap[prop];
                var newIdArr = [];
                _.map(arr, function (item, i) {
                    newIdArr.push(item.gid);
                });
                //文件个数
                fileLength = fileLength + newIdArr.length;
                map[prop] = newIdArr.join(',');
            }
        }
        /*var idArr = [];
        this.state.data.map(function(item,i){
            idArr.push(item.gid);
        })
        this.props.modal.changeValue(idArr.join(','));*/

        this.props.modal.changeValue(JSON.stringify(map), fileLength);
    },
    staticUrl: function staticUrl() {
        if (this.props.fileUrl) {
            return this.props.fileUrl;
        }
        return this.props.serverUrl;
    },
    timestampToDate: function timestampToDate(ts) {
        return new Date(parseInt(ts)).toLocaleString().replace(/:\d{1,2}$/, ' ');
    },
    render: function render() {
        var that = this;
        var totalsize = 0;
        var data = this.state.data ? this.state.data : [];
        var sourceId = this.props.sourceId;
        var billType = this.props.billType;
        var sourceType = this.props.sourceType;
        var visible = this.props.pageVisible;
        var uploadFileUrl = this.props.serverUrl + uploadUrl;
        var downloadFileUrl = this.props.serverUrl + downloadUrl;
        var staticUrl = this.staticUrl();
        //console.log("sourceId:"+sourceId+",billType:"+billType+",sourceType:"+sourceType);
        //附件上传
        var uploadProps = {
            action: uploadFileUrl,
            multiple: false,
            showUploadList: true,
            data: { sourceId: sourceId, billType: billType, sourceType: sourceType, userId: that.state.userId, userName: that.state.userName },
            beforeUpload: function beforeUpload(info) {
                if (info.size == 0) {
                    showErrMsg('上传失败', '请不要上传空文件');
                    return false;
                }
                //上传前判断
                if (that.props.beforeUpload) {
                    if (!that.props.beforeUpload()) {
                        return false;
                    }
                }
                that.state.loadingDisplay = true;
                that.setState({ changeFlag: !that.state.changeFlag });
            },
            onChange: function onChange(info) {
                var files = info.fileList;
                if (info.file.status === 'done') {
                    var resultObj = info.file.response;
                    if (resultObj.success) {
                        var result = that.state.data;
                        if (result) {
                            result = resultObj.backData.concat(result);
                        } else {
                            result = resultObj.backData;
                        }
                        that.setState({ data: result });
                        that.state.dataMap[sourceType] = result;
                        that.onChangePage(1);

                        that.refreshValue();
                    } else {
                        showErrMsg('上传失败', info.file.response.msg);
                    }
                    that.state.loadingDisplay = false;
                    that.setState({ changeFlag: !that.state.changeFlag });
                    that.setState({ uploadEvent: null, fileList: [] });
                    files = [];
                } else if (info.file.status === 'error') {
                    showErrMsg('上传失败', '请求失败');
                    that.state.loadingDisplay = false;
                    that.setState({ changeFlag: !that.state.changeFlag });
                    that.setState({ uploadEvent: null, fileList: [] });
                    files = [];
                }
                that.setState({ uploadEvent: info.event, fileList: files });
            }
        };
        var pagination = {
            total: data.length,
            showSizeChanger: true,
            onShowSizeChange: function onShowSizeChange(current, pageSize) {}
        };
        var columns = [{
            title: '序号',
            dataIndex: 'num',
            isShow: true,
            width: 60,
            key: 'num',
            render: function render(text, record, index) {
                return React.createElement(
                    'span',
                    null,
                    index + 1
                );
            }
        }, {
            title: '名称',
            isShow: true,
            dataIndex: 'fileName',
            key: 'fileName',
            render: function render(text, record, index) {
                if (record.gid) {
                    //let urlName = downloadFileUrl+record.gid;
                    var urlName = handleUrl(staticUrl + "/" + record.filePath + "?filename=" + record.fileName);
                    return React.createElement(
                        'a',
                        { className: 'title-href', target: '_blank', href: urlName, title: record.fileName },
                        record.fileName
                    );
                }
                return "";
            }
        }, {
            title: '上传人',
            isShow: true,
            width: 90,
            dataIndex: 'createName',
            key: 'createName'
        }, {
            title: '类型',
            isShow: true,
            dataIndex: 'docType',
            width: 80,
            key: 'docType',
            render: function render(text, record, index) {
                if (record.fileName) {
                    var fileName = record.fileName;
                    var idx = fileName.lastIndexOf('.');
                    if (idx > -1) {
                        return fileName.substr(idx + 1);
                    }
                }
                return "";
            }
        }, {
            title: '大小(KB)',
            isShow: true,
            width: 80,
            dataIndex: 'fileSize',
            key: 'fileSize',
            render: function render(text, record, index) {
                if (record.fileSize) {
                    return execute(record.fileSize, 1024, 3, 1);
                }
                return "";
            }
        }, {
            title: '上传时间',
            isShow: true,
            dataIndex: 'createDate',
            width: 80,
            key: 'createDate',
            render: function render(text, record, index) {
                if (record.createDate) {
                    return formatDate(new Date(parseInt(record.createDate)), "yyyy-MM-dd");
                }
                return "";
            }
        }];
        var rowSelection = {
            onChange: function onChange(selectedRowKeys, selectedRows) {
                that.setState({ selectId: selectedRowKeys });
                that.setState({ selectItems: selectedRows });
            },
            onSelect: function onSelect(record, selected, selectedRows) {},
            onSelectAll: function onSelectAll(selected, selectedRows, changeRows) {},

            selectedRowKeys: this.state.selectId
        };

        var showUploadArea = this.props.showUploadBtn || this.props.showDelBtn;
        var showUpload = this.props.showUploadBtn;
        var showDel = this.props.showDelBtn;
        var readOnly = this.props.readOnly;
        var uploadProgress = this.state.uploadEvent;
        var fileList = this.state.fileList;
        return React.createElement(
            'div',
            { className: 'attachment-page' },
            showUpload || showDel ? React.createElement(
                YYToolbar,
                { className: 'attachment-toolbar' },
                showUpload && !readOnly ? React.createElement(
                    YYUpload,
                    _extends({}, uploadProps, { fileList: fileList }),
                    React.createElement(
                        YYButton,
                        { type: 'primary', ghost: true, onClick: that.onUpload },
                        '\u9644\u4EF6\u4E0A\u4F20'
                    )
                ) : null,
                showDel && !readOnly ? React.createElement(
                    YYButton,
                    { type: 'error', ghost: true, onClick: this.deleteAll },
                    '\u6279\u91CF\u5220\u9664'
                ) : null
            ) : null,
            React.createElement(
                'div',
                { className: 'attachmentmgr' },
                React.createElement(YYTable, { rowKey: function rowKey(record) {
                        return record.gid;
                    }, dataSource: data, pagination: pagination,
                    columns: columns, rowSelection: rowSelection })
            )
        );
    }
});
module.exports = AttachMgrTable;
