// yylib-react/dist/business/attach/mgr/AttachMgrModal.js
'use strict';

/**
 * Created by Administrator on 2016/7/26.
 */
var React = require('react');
var AttachTable = require('./AttachMgrTable');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass,
    YYModal = _require.YYModal,
    YYButton = _require.YYButton,
    YYTree = _require.YYTree,
    YYTreeNode = _require.YYTreeNode,
    YYCol = _require.YYCol,
    YYRow = _require.YYRow;

var _require2 = require('lodash'),
    isFunction = _require2.isFunction;

var ADDR = require('../../BaseHost');
var AttachmentMgrModal = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        return { serverUrl: ADDR, fileUrl: ADDR, value: null, fileTypes: '', showUploadBtn: true, showDelBtn: true,
            readOnly: false, beforeUpload: null, beforeDel: null, userId: '', userName: '', enableUser: true };
    },
    getInitialState: function getInitialState() {
        return {
            visible: false,
            sourceId: this.props.sourceId || null,
            billType: null,
            sourceType: this.props.sourceType ? this.props.sourceType : '102345678',
            loading: false,
            value: null,
            defaultKey: ''
        };
    },

    changeValue: function changeValue(value, fileLength) {
        if (!this.props.readOnly) {
            if (isFunction(this.props.onChange)) {
                this.props.onChange(value, fileLength);
            }
        }
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        //解决sourceId发生变化的时候组件的返回值问题
        var lastSourceId = this.state.sourceId;
        if (!lastSourceId) {
            lastSourceId = "";
        }
        var sourceId = nextProps.sourceId;
        if (!sourceId) {
            sourceId = "";
        }
        this.setState({ sourceId: sourceId });
        if (sourceId != lastSourceId) {
            //如果sourceId变化，则组件的值为空
            this.changeValue(null);
        }
        if (nextProps.sourceType) {
            this.setState({ sourceType: nextProps.sourceType });
        }
    },
    componentWillMount: function componentWillMount() {
        var fileTypeMap = this.props.fileTypes;
        if (fileTypeMap) {
            var defaultKey = fileTypeMap[0].code;
            this.setState({ defaultKey: defaultKey, fileTypes: fileTypeMap, sourceType: defaultKey });
        }
    },
    handleCancel: function handleCancel() {
        if (isFunction(this.props.onCancel)) {
            this.props.onCancel();
        }
    },
    onSelectNode: function onSelectNode(selectedKeys, info) {
        //console.log(selectedKeys,info);
        var key = selectedKeys[0];
        this.setState({ sourceType: key });
    },
    render: function render() {
        var that = this;
        var nameMap = this.props.fileTypes;
        //是否定义文件分类
        var showTree = false;
        var tableWidth = 24;
        if (nameMap) {
            showTree = true;
            tableWidth = 19;
        }
        var defaultKey = this.state.defaultKey;
        var loop = function loop(data) {
            return data.map(function (item) {
                return React.createElement(YYTreeNode, { key: item.code, title: item.name });
            });
        };
        return React.createElement(
            YYModal,
            { ref: 'modal',
                wrapClassName: 'vertical-center-infomodal',
                visible: this.props.visible,
                title: '\u9644\u4EF6\u7BA1\u7406', width: '800',
                onCancel: this.handleCancel, footer: [React.createElement(
                    YYButton,
                    { key: 'back', onClick: this.handleCancel },
                    '\u5173\u95ED'
                )] },
            React.createElement(
                YYRow,
                { className: 'batch-opt' },
                showTree ? React.createElement(
                    YYCol,
                    { span: '5' },
                    React.createElement(
                        YYTree,
                        { defaultSelectedKeys: defaultKey, onSelect: this.onSelectNode },
                        loop(nameMap)
                    )
                ) : '',
                React.createElement(
                    YYCol,
                    { span: tableWidth },
                    React.createElement(AttachTable, { serverUrl: this.props.serverUrl, fileUrl: this.props.fileUrl, readOnly: this.props.readOnly, modal: this, billType: this.props.billType,
                        showUploadBtn: this.props.showUploadBtn, showDelBtn: this.props.showDelBtn, sourceId: this.state.sourceId,
                        sourceType: this.state.sourceType, pageVisible: true, beforeUpload: this.props.beforeUpload,
                        beforeDel: this.props.beforeDel, userId: this.props.userId, userName: this.props.userName, enableUser: this.props.enableUser })
                )
            )
        );
    }
});
module.exports = AttachmentMgrModal;
