// yylib-react/dist/yylib2/refer/constants/ReferConstants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
//参照组件用常量

//请求url
var GET_REFINFO_BYCODE = exports.GET_REFINFO_BYCODE = '/icop-support-web/refer/findByCode';
// url: 查询常用
// export const GET_REFER_OFTEN = '/icop-support-web/store/showlist?userId={userId}&refinfokey={refinfokey}&conditionly&searchtext';
var GET_REFER_OFTEN = exports.GET_REFER_OFTEN = '/icop-support-web/store/showlist?userId={userId}&refinfokey={refinfokey}&conditionly&searchtext={searchtext}';
// url：保存常用
var SAVE_REFER_OFTEN = exports.SAVE_REFER_OFTEN = '/icop-support-web/store/save';
// url: 删除常用
var DELETE_REFER_OFTEN = exports.DELETE_REFER_OFTEN = '/icop-support-web/store/delete?userId={userId}&refinfokey={refinfokey}&conditionly&ids={ids}';
//获取参照配置表
var GET_REFINFO_BYCODE_SUCCESS = exports.GET_REFINFO_BYCODE_SUCCESS = 'GET_REFINFO_BYCODE_SUCCESS';
var GET_REFINFO_BYCODE_ERROR = exports.GET_REFINFO_BYCODE_ERROR = 'GET_REFINFO_BYCODE_ERROR';

//获取参照树数据
var GET_TREEDATA_REF_SUCCESS = exports.GET_TREEDATA_REF_SUCCESS = 'GET_TREEDATA_REF_SUCCESS';
var GET_TREEDATA_REF_ERROR = exports.GET_TREEDATA_REF_ERROR = 'GET_TREEDATA_REF_ERROR';
//获取参照表数据
var GET_GRIDDATA_REF_SUCCESS = exports.GET_GRIDDATA_REF_SUCCESS = 'GET_GRIDDATA_REF_SUCCESS';
var GET_GRIDDATA_REF_ERROR = exports.GET_GRIDDATA_REF_ERROR = 'GET_GRIDDATA_REF_ERROR';
//获取参照树表 grid
var GET_TREEGRIDDATA_REF_SUCCESS = exports.GET_TREEGRIDDATA_REF_SUCCESS = 'GET_TREEGRIDDATA_REF_SUCCESS';
var GET_TREEGRIDDATA_REF_ERROR = exports.GET_TREEGRIDDATA_REF_ERROR = 'GET_TREEGRIDDATA_REF_ERROR';
// 获取参照树表： tree & grid
var GET_TREE_AND_GRID_REF_SUCCESS = exports.GET_TREE_AND_GRID_REF_SUCCESS = 'GET_TREE_AND_GRID_REF_SUCCESS';

// 查询常用成功
var GET_REFER_OFTEN_SUCCESS = exports.GET_REFER_OFTEN_SUCCESS = 'GET_REFER_OFTEN_SUCCESS';
// 查询常用失败
var GET_REFER_OFTEN_ERROR = exports.GET_REFER_OFTEN_ERROR = 'GET_REFER_OFTEN_ERROR';
// 保存常用成功
var SAVE_REFER_OFTEN_SUCCESS = exports.SAVE_REFER_OFTEN_SUCCESS = 'SAVE_REFER_OFTEN_SUCCESS';
// 保存常用失败
var SAVE_REFER_OFTEN_ERROR = exports.SAVE_REFER_OFTEN_ERROR = 'SAVE_REFER_OFTEN_ERROR';
// 删除常用成功
var DELETE_REFER_OFTEN_SUCCESS = exports.DELETE_REFER_OFTEN_SUCCESS = 'DELETE_REFER_OFTEN_SUCCESS';
// 删除常用失败
var DELETE_REFER_OFTEN_ERROR = exports.DELETE_REFER_OFTEN_ERROR = 'DELETE_REFER_OFTEN_ERROR';

// 重加载参照数据
var DO_RELOAD_REF = exports.DO_RELOAD_REF = 'DO_RELOAD_REF';
// 清空参照数据
var DO_CLEAR_PREDATAS = exports.DO_CLEAR_PREDATAS = 'DO_CLEAR_PREDATAS';
// 参照类型(用于在action中判断树查询、表查询是否来自树表): 树表
var REFER_TYPE_TREEGRID = exports.REFER_TYPE_TREEGRID = 'treegrid';

//storeKey前缀： 常用数据
var REFER_OFTEN = exports.REFER_OFTEN = 'refer-often-';
//storeKey前缀： keyField
var REFER_KEYFIELD = exports.REFER_KEYFIELD = 'refer-keyfield-';
//storeKey前缀： keyField
var REFER_TAGFIELD = exports.REFER_TAGFIELD = 'refer-tagfield-';
//storeKey前缀： 树表中选择的treeurl(树切换场景)
var REFER_SELECTED_TREEURL = exports.REFER_SELECTED_TREEURL = 'refer-selected-treeurl-';
// tablekey前缀
var REFER_TABLEKEY = exports.REFER_TABLEKEY = 'refer-tablekey-';
// storeKey前缀: 树表中选择的nodeKey(树表打开默认select上次select的树节点id)
var REFER_TREEGRID_SELECTED_NODEKEY = exports.REFER_TREEGRID_SELECTED_NODEKEY = 'refer-treegrid-selected-nodekey-';
// storeKey前缀: 树表中选择的nodeKey(树表打开默认select上次select的树节点)
var REFER_TREEGRID_SELECTED_NODE = exports.REFER_TREEGRID_SELECTED_NODE = 'refer-treegrid-selected-node-';

// 常用字符串：添加常用
var REFER_ADD_OFTEN = exports.REFER_ADD_OFTEN = '添加常用';
// 常用字符串：删除常用
var REFER_DELETE_OFTEN = exports.REFER_DELETE_OFTEN = '删除常用';
// 参照表格列默认宽度：200
var REFER_DEFAULT_COLUMN_WIDTH = exports.REFER_DEFAULT_COLUMN_WIDTH = 200;
// 参照表格操作列默认宽度; 65
var REFER_OPERATE_COLUMN_WIDTH = exports.REFER_OPERATE_COLUMN_WIDTH = 100;
// 参照表格操作列默认宽度; 65
var REFER_CHECKBOX_WIDTH = exports.REFER_CHECKBOX_WIDTH = 30;
// 参照表格操作列默认宽度; 65
var REFER_TABLE_PAGINATION_HEIGHT = exports.REFER_TABLE_PAGINATION_HEIGHT = 30;

// 与后台约定的参数名称：用于打开参照中的超链接，传参数
var REFER_PARAM_RELYFIELD = exports.REFER_PARAM_RELYFIELD = 'relyfield';
// 与后台约定的参数名称：选择的树节点，用于打开参照中的超链接，传参数
var REFER_PARAM_SELECTEDNODEID = exports.REFER_PARAM_SELECTEDNODEID = 'selectedNodeId';

var REFER_ERROR_MSG = exports.REFER_ERROR_MSG = '调用服务失败';

var LOADING_REFER_TREE = exports.LOADING_REFER_TREE = 'LOADING_REFER_TREE';

var LOADING_REFER_GRID = exports.LOADING_REFER_GRID = 'LOADING_REFER_GRID';
