// yylib-react/dist/yylib2/index.js
'use strict';

//antd.min.css直接在页面引入：<link href="http://cdn.bootcss.com/antd/1.11.1/antd.min.css" rel="stylesheet">
//require('antd.css');
require('./style/common.less');
//类构建器
var YYClass = require('./base/YYClass');
//读取版本信息
require('./version');
var yylibui = {
  YYClass: YYClass,
  //容器类
  YYApp: require('./container/YYApp'),
  YYPage: require('./container/YYPage'),
  YYArea: require('./container/YYArea'),
  YYRow: require('./layout/YYRow'),
  YYCol: require('./layout/YYCol'),
  YYToolbar: require('./toolbar/YYToolbar'),
  YYPanel: require('./panel/YYPanel'),
  YYScrollbar: require('./scrollbar/YYScrollbar'),
  YYTabs: require('./tabs/YYTabs'),
  YYTab: require('./tabs/YYTab'),
  YYAccordions: require('./accordions/YYAccordions'),
  // YYAccordion : YYAccordions.Panel,
  YYAccordion: require('./accordions/YYAccordion'),
  YYPagingButton: require('./paging/YYPagingButton'),
  YYPagination: require('./paging/YYPagination'),
  YYBreadcrumb: require('./breadcrumb/YYBreadcrumb'),
  YYBreadcrumbItem: require('./breadcrumb/YYBreadcrumb').Item,
  YYSteps: require('./steps/YYSteps'),
  YYStep: require('./steps/YYStep'),
  YYNavItem: require('./steps/YYNavItem'),
  YYSplitPane: require('./splitpane/YYSplitPane'),
  //  YYStep : YYSteps.Step,

  //基础类
  YYIcon: require('./icon/YYIcon'),
  YYButton: require('./button/YYButton'),
  YYButtonGroup: require('./button/YYButton').Group,
  YYButtonSelect: require('./button/YYButtonSelect'),
  YYButtonItem: require('./button/YYButtonItem'),

  YYMenu: require('./menu/YYMenu'),
  YYMenuDivider: require('./menu/YYMenuDivider'),
  YYMenuGroup: require('./menu/YYMenuGroup'),
  YYMenuSub: require('./menu/YYMenuSub'),
  YYMenuItem: require('./menu/YYMenuItem'),

  //数据类
  YYMenuList: require('./menulist/YYMenuList'),
  YYTransfer: require('./transfer/YYTransfer'),
  YYTree: require('./tree/YYTree'),
  YYTreeNode: require('./tree/YYTreeNode'),
  YYReferDialog: require('./refer/YYReferDialog'),
  YYReferInput: require('./refer/YYReferInput'),
  YYTable: require('./table/YYTable'),
  YYTableCol: require('./table/YYTableCol'),
  YYColumnConfig: require('./table/YYColumnConfig'),
  YYEditTable: require('./table/YYEditTable'),
  YYCount: require('./table/YYCount'), // 表格统计组件
  YYCalc: require('./calc/YYCalc'),
  YYCalcInput: require('./calc/YYCalcInput'),

  // YYChart : require('./chart/YYChart'),
  // 特殊类
  YYHighLight: require('./highlight/YYHighLight'),
  YYDraggable: require('./draggable/YYDraggable'),
  YYAlert: require('./alert/YYAlert'),
  YYBadge: require('./badge/YYBadge'),
  YYCalendar: require('./calendar/YYCalendar'),
  YYCard: require('./card/YYCard'),
  YYCarousel: require('./carousel/YYCarousel'),
  YYDropdown: require('./dropdown/YYDropdown'),
  YYDropdownButton: require('./dropdown/YYDropdownButton'),
  YYDivide: require('./divide/YYDivide'),
  // YYDropdownButton : YYDropdown.Button,
  YYModal: require('./modal/YYModal'),
  YYPopconfirm: require('./popconfirm/YYPopconfirm'),
  YYPopover: require('./popover/YYPopover'),
  YYProgress: require('./progress/YYProgress'),
  YYTag: require('./tag/YYTag'),
  YYTimeline: require('./timeline/YYTimeline'),
  // YYTimelineItem : YYTimeline.Item,
  YYTimelineItem: require('./timeline/YYTimelineItem'),
  YYTooltip: require('./tooltip/YYTooltip'),
  YYColorPicker: require('./colorpicker/YYColorPicker'),
  YYOrgchart: require('./orgchart/YYOrgchart'),
  //表单类
  YYCascader: require('./formcontrols/cascader/YYCascader'),
  YYCheckbox: require('./formcontrols/checkbox/YYCheckbox'),
  YYCheckboxGroup: require('./formcontrols/checkbox/YYCheckboxGroup'),
  YYDatePicker: require('./formcontrols/datepicker/YYDatePicker'),
  YYMonthPicker: require('./formcontrols/datepicker/YYMonthPicker'),
  YYYearPicker: require('./formcontrols/datepicker/YYYearPicker'),
  YYRangePicker: require('./formcontrols/datepicker/YYRangePicker'),
  YYForm: require('./formcontrols/form/YYForm'),
  YYFormItem: require('./formcontrols/form/YYFormItem'),
  YYInput: require('./formcontrols/input/YYInput'),
  YYTextarea: require('./formcontrols/textarea/YYTextarea'),
  YYInputGroup: require('./formcontrols/input/YYInputGroup'),
  YYInputButton: require('./formcontrols/inputbutton/YYInputButton'),
  YYInputNumber: require('./formcontrols/inputnumber/YYInputNumber'),
  YYInputCurrency: require('./formcontrols/inputnumber/YYInputCurrency'),
  YYRadio: require('./formcontrols/radio/YYRadio'),
  YYRadioButton: require('./formcontrols/radio/YYRadioButton'),
  YYRadioGroup: require('./formcontrols/radio/YYRadioGroup'),
  YYRate: require('./formcontrols/rate/YYRate'),
  YYSelect: require('./formcontrols/select/YYSelect'),
  YYOption: require('./formcontrols/select/YYOption'),
  YYOptGroup: require('./formcontrols/select/YYSelect').OptGroup,

  YYSlider: require('./formcontrols/slider/YYSlider'),
  YYSwitch: require('./formcontrols/switch/YYSwitch'),
  YYTimePicker: require('./formcontrols/timepicker/YYTimePicker'),
  YYUpload: require('./formcontrols/upload/YYUpload'),
  YYUploadDragger: require('./formcontrols/upload/YYUploadDragger'),
  YYTreeSelect: require('./formcontrols/treeselect/YYTreeSelect'),

  YYTagItem: require('./formcontrols/tag/YYTagItem'),
  YYTagGroup: require('./formcontrols/tag/YYTagGroup'),

  YYSpin: require('./spin/YYSpin'),
  YYBacktop: require('./backtop/YYBacktop'),
  YYAffix: require('./affix/YYAffix'),
  YYQueueanim: require('./queueanim/YYQueueanim'),
  YYIframe: require('./iframe/YYIframe'),
  //  YYEditor : require('./editor/YYEditor'),

  //工具类
  YYMessage: require('./message/YYMessage'),
  YYNotice: require('./notification/YYNotification'),
  YYClipboard: require('./clipboard/YYClipboard'),
  YYAssistant: require('./assistant/YYAssistant'),
  YYPopMenu: require('./popmenu/YYPopMenu'),

  YYTemplate: require('./template/YYTemplate'),
  YYStateTrack: require('./statetrack/YYStateTrack'),
  YYInputTag: require('./inputtag/YYInputTag')
  // YYTile : require('./tiles/YYTile'),
  // YYTiles : require('./tiles/YYTiles'),
  // YYTilesGroup : require('./tiles/YYTilesGroup'),

};

module.exports = yylibui;
