// yylib-react/dist/utils/asyncLoader.js
'use strict';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 异步的加载组件
 */
var asyncLoader = function asyncLoader(component) {
    return _react2.default.createClass({
        getInitialState: function getInitialState() {
            return {
                Component: null
            };
        },
        componentDidMount: function componentDidMount() {
            var _this = this;

            //模拟出loading的效果
            // setTimeout(() => {
            //   component((Component) => {
            //     this.setState({
            //       Component: Component
            //     });
            //   });
            // }, 3000);
            component(function (Component) {
                _this.setState({
                    Component: Component
                });
            });
        },
        render: function render() {
            var Component = this.state.Component;
            if (Component) {
                return _react2.default.createElement(Component, this.props);
            } else {
                //return <div>Loading...</div>;
                return _react2.default.createElement('div', null);
            }
        }
    });
};
module.exports = asyncLoader;
