// yylib-react/dist/utils/PermissionToken.js
'use strict';

var ajax = require('./ajax');

var PermissionToken = {
    PERMISSIONCENTERBASEHOST: 'http://60.205.12.0:81',
    MOD_PUBBASE_BTNGROUP_PERMISSION: '/icop-permission-web/',
    BTN_S_PERMISSOIN: 'roleAppBtn/btnsPermission/',

    getBtnGroupPermission: function getBtnGroupPermission(userId, appId, _success, fail) {
        var self = this;
        ajax.request({
            method: 'GET',
            url: self.PERMISSIONCENTERBASEHOST + self.MOD_PUBBASE_BTNGROUP_PERMISSION + self.BTN_S_PERMISSOIN + userId + '/' + appId,
            type: 'json',
            success: function success(data) {
                if (data.code == "SUCCESS") {
                    _success(data.msg, data.data);
                } else {
                    fail(data.msg);
                }
            },
            error: function error(data) {
                fail('网络异常或者内部错误!');
            }
        });
    },

    setPermissionCenterBaseHost: function setPermissionCenterBaseHost(url) {
        this.PERMISSIONCENTERBASEHOST = url;
    }
};

module.exports = PermissionToken;
