// yylib-react/dist/yylib2/transfer/YYTransferSearch.js
'use strict';

/**
 * Created by wuhao on 16/7/19.
 */

var React = require('react');
var YYClass = require('../base/YYClass');
var YYIcon = require('../icon/YYIcon');
var YYInput = require('../formcontrols/input/YYInput');
require('./style/index.less');

function noop() {}

var YYTransferSearch = YYClass.create({
    uiName: 'transfer-search',
    getDefaultProps: function getDefaultProps() {
        return {
            searchPlaceholder: '',
            handleOnInputChange: noop,
            handleClear: noop,
            handleFindPrevious: noop,
            handleFindNext: noop
        };
    },
    handleOnInputChange: function handleOnInputChange(e) {
        this.props.handleOnInputChange(e);
    },
    handleClear: function handleClear(e) {
        e.preventDefault();
        this.props.handleClear(e);
    },
    handleFindPrevious: function handleFindPrevious(e) {
        this.props.handleFindPrevious(e);
    },
    handleFindNext: function handleFindNext(e) {
        this.props.handleFindNext(e);
    },
    render: function render() {
        var value = this.props.value;

        return React.createElement(
            'div',
            { className: this.className() },
            React.createElement(YYInput, { placeholder: this.props.searchPlaceholder,
                onClearClick: this.handleClear,
                onPressEnter: this.handleFindNext,
                onChange: this.handleOnInputChange,
                value: value
            }),
            React.createElement(YYIcon, { type: 'down', className: 'search-next', title: '\u67E5\u4E0B\u4E00\u4E2A', onClick: this.handleFindNext }),
            React.createElement(YYIcon, { type: 'up', className: 'search-previous', title: '\u67E5\u4E0A\u4E00\u4E2A', onClick: this.handleFindPrevious })
        );
    }
});

module.exports = YYTransferSearch;
