// yylib-react/dist/yylib2/calc/YYCalcButton.js
'use strict';

/**
 * 计算器按钮组件
 */
var React = require('react');
var ReactDoM = require('react-dom');
var YYClass = require('../base/YYClass');
var field = YYClass.field;
var YYClass = require('../base/YYClass');
var PropTypes = YYClass.PropTypes;
//定义META
var uiMeta = {
    name: 'calcbutton',
    fields: [
        //field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    ]
};
var YYcalcbutton = YYClass.create({
    uiMeta: uiMeta,
    keyMapping: {},
    onClick: function onClick(event) {
        var target = event.target || event.srcElement;
        target.classList.remove('clicked');
        setTimeout(function () {
            target.classList.add('clicked');
        }, 0);
        this.props.onClick(target.getAttribute('data-value'));
    },
    componentDidMount: function componentDidMount() {
        var _this = this;

        var dom = ReactDoM.findDOMNode(this);
        var buttons = dom.querySelectorAll('button');
        buttons = [].slice.call(buttons);
        buttons.forEach(function (button) {
            _this.keyMapping[button.getAttribute('data-code')] = button;
        });
        window.onkeydown = function (event) {
            var button;
            var key = (event.shiftKey ? 'shift+' : '') + event.keyCode || event.which;
            if (button = _this.keyMapping[key]) {
                button.click();
                event.stopPropagation();
            }
        };
    },
    render: function render() {
        var rowcls = this.className('row');
        var colcls = this.className('col');
        var btncls = this.className('button');
        return React.createElement(
            'div',
            { className: this.className() + ' ' + rowcls },
            React.createElement(
                'div',
                { className: colcls + ' calcs3' },
                React.createElement(
                    'div',
                    { className: rowcls + ' calcs1' },
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '67', 'data-value': 'c', onClick: this.onClick },
                        'C'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '8', 'data-value': 'back', onClick: this.onClick },
                        '\u2190'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '191', 'data-value': '/', onClick: this.onClick },
                        '\xF7'
                    )
                ),
                React.createElement(
                    'div',
                    { className: rowcls + ' calcs1' },
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '55', 'data-value': '7', onClick: this.onClick },
                        '7'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '56', 'data-value': '8', onClick: this.onClick },
                        '8'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '57', 'data-value': '9', onClick: this.onClick },
                        '9'
                    )
                ),
                React.createElement(
                    'div',
                    { className: rowcls + ' calcs1' },
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '52', 'data-value': '4', onClick: this.onClick },
                        '4'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '53', 'data-value': '5', onClick: this.onClick },
                        '5'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '54', 'data-value': '6', onClick: this.onClick },
                        '6'
                    )
                ),
                React.createElement(
                    'div',
                    { className: rowcls + ' calcs1' },
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '49', 'data-value': '1', onClick: this.onClick },
                        '1'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '50', 'data-value': '2', onClick: this.onClick },
                        '2'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '51', 'data-value': '3', onClick: this.onClick },
                        '3'
                    )
                ),
                React.createElement(
                    'div',
                    { className: rowcls + ' calcs1' },
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs2', 'data-code': '48', 'data-value': '0', onClick: this.onClick },
                        '0'
                    ),
                    React.createElement(
                        'button',
                        { className: btncls + ' calcs1', 'data-code': '190', 'data-value': '.', onClick: this.onClick },
                        '.'
                    )
                )
            ),
            React.createElement(
                'div',
                { className: colcls + ' calcs1' },
                React.createElement(
                    'button',
                    { className: btncls + ' calcs1', 'data-code': 'shift+56', 'data-value': '*', onClick: this.onClick },
                    '\xD7'
                ),
                React.createElement(
                    'button',
                    { className: btncls + ' calcs1', 'data-code': '189', 'data-value': '-', onClick: this.onClick },
                    '-'
                ),
                React.createElement(
                    'button',
                    { className: btncls + ' calcs1', 'data-code': '187', 'data-value': '+', onClick: this.onClick },
                    '+'
                ),
                React.createElement(
                    'button',
                    { className: btncls + ' calcs2' + ' button-equal', 'data-code': '13', 'data-value': '=', onClick: this.onClick },
                    '='
                )
            )
        );
    }
});
module.exports = YYcalcbutton;
