// yylib-react/dist/yylib2/calc/YYCalcInput.js
'use strict';

var React = require('react');

var _require = require('react-dom'),
    findDOMNode = _require.findDOMNode;

var Trigger = require('rc-trigger');
var YYClass = require('../base/YYClass');
var YYInput = require('../formcontrols/input/YYInput');
var YYIcon = require('../icon/YYIcon');
var YYCalc = require('./YYCalc');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
require('./style/index2.less');
//定义META
var uiMeta = {
    name: 'calcinput',
    fields: [field('size', 'default', PropTypes.string, false, '设置input输入框大小，给定三种尺寸,small,default,large')]
};
var builtinPlacements = {
    bottom: {
        points: ['tc', 'bc']
    }
};
var YYCalcInput = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        return {
            value: '',
            placement: 'bottom'
        };
    },
    changeValue: function changeValue(event) {
        var _state = this.refs.yycalc.state;
        var value = _state.result || _state.formula + _state.opr + _state.input_num;
        this.setState({
            value: value
        });
    },
    onEqual: function onEqual(result) {
        this.setState({
            value: result
        });
    },
    preventDefault: function preventDefault(e) {
        e.preventDefault();
    },
    render: function render() {
        var _this = this;

        return React.createElement(
            'div',
            { className: this.className(), ref: 'YYCalcInput' },
            React.createElement(
                Trigger,
                { popup: React.createElement(YYCalc, { onEqual: this.onEqual, ref: 'yycalc' }),
                    action: ['click'],
                    getPopupContainer: function getPopupContainer() {
                        return _this.refs.YYCalcInput;
                    },
                    popupPlacement: this.state.placement,
                    builtinPlacements: builtinPlacements
                },
                React.createElement(
                    'div',
                    { className: 'calcinput-area' },
                    React.createElement(YYInput, { onChange: this.changeValue,
                        value: this.state.value,
                        onClick: this.preventDefault,
                        size: this.props.size
                    }),
                    React.createElement(YYIcon, { type: 'calculator' })
                )
            )
        );
    }
});
module.exports = YYCalcInput;
