// yylib-react/dist/yylib2/colorpicker/YYColorPicker.js
'use strict';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

/**
 * Created by wuhao on 16/7/21.
 */

var React = require('react');

var _require = require('react-dom'),
    findDOMNode = _require.findDOMNode;

var YYClass = require('../base/YYClass');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
var YYToolbar = require('../toolbar/YYToolbar');
var YYMessage = require('../message/YYMessage');
var YYButton = require('../button/YYButton');
var ColorPicker = require('rc-color-picker');
require('rc-color-picker/assets/index.css');
require('./style/index.less');
function noop() {}
var uiMeta = {
    name: 'colorpicker',
    fields: [field('animation', 'slide-up', PropTypes.string, false, '打开面板的动画'), field('transitionName', '', PropTypes.number, false, '动画类名'), field('getCalendarContainer', function () {
        return document.body;
    }, PropTypes.func, false, '组件父容器'), field('align', {}, PropTypes.object, false, '组件放置参数'), field('alpha', 100, PropTypes.number, false, '初始透明度'), field('defaultAlpha', 100, PropTypes.number, false, '初始默认透明度'), field('color', '#36c', PropTypes.string, false, '初始颜色'), field('defaultColor', '#36c', PropTypes.string, false, '初始默认颜色'), field('onChange', noop, PropTypes.func, false, '颜色改变时的回调'), field('onOpen', noop, PropTypes.func, false, '取色面板打开时的回调'), field('onClose', noop, PropTypes.func, false, '取色面板关闭时的回调'), field('placement', 'topLeft', PropTypes.oneOf(['topLeft', 'topRight', 'bottomLeft', 'bottomRight']), false, '相对触发元素的位置'), field('mode', 'RGB', PropTypes.oneOf(['RGB', 'HSB', 'HSL']), false, '颜色显示模式')]
};
var YYColorPicker = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        var value = this.props.value;
        return {
            isShow: false,
            defaultValue: value,
            value: value
        };
    },
    statics: {
        Panel: ColorPicker.Panel
    },
    _isValueRight: function _isValueRight(value) {
        var values = [].concat(_toConsumableArray(_.trim(value)));
        if (values[0] == '#') {
            values.shift();
        } else {
            return false;
        }
        if (values.length != 3 && values.length != 6) return false;
        var flag = values.every(function (val, index, arr) {
            var code = val.charCodeAt(0);
            return 48 <= code && code <= 57 || 65 <= code && code <= 70 || 97 <= code && code <= 102;
        });
        return flag;
    },
    showPanel: function showPanel() {
        this.setState({
            isShow: !this.state.isShow
        });
        // if(this._isValueRight(_.clone(value))){
        //     this.setState({
        //         isShow:!this.state.isShow
        //     })
        // }else{
        //     YYMessage.warn('请输入正确颜色值');
        // }
    },
    onChange: function onChange(obj) {
        this.setState({
            defaultValue: obj.color
        });
    },
    _clickOk: function _clickOk() {
        var defaultValue = this.state.defaultValue;
        this.setState({
            value: defaultValue,
            isShow: false
        });
        if (typeof this.props.onChange == 'function') {
            this.props.onChange(defaultValue);
        }
    },
    componentDidMount: function componentDidMount() {
        this._addEvent();
    },
    _addEvent: function _addEvent() {
        document.addEventListener('mousedown', this._onDocumentClick);
    },
    componentWillUnmount: function componentWillUnmount() {
        document.removeEventListener('mousedown', this._onDocumentClick);
    },
    _onDocumentClick: function _onDocumentClick(e) {
        var isShow = this.state.isShow;
        var target = e.target;
        if (isShow) {
            var colorPanel = findDOMNode(this.refs.panel).parentNode;
            var trigger = findDOMNode(this).querySelector('.colortrigger');
            if (target != trigger && !colorPanel.contains(target)) {
                this.setState({
                    isShow: false
                });
            }
        }
    },
    _getPosition: function _getPosition() {
        var trigger = findDOMNode(this).querySelector('.colortrigger');
        if (trigger) {
            var trigger = trigger.getBoundingClientRect();
            return trigger;
        }
    },
    _clickCancel: function _clickCancel() {
        this.setState({
            isShow: false
        });
    },
    _clickReset: function _clickReset() {
        this.setState({
            value: this.props.value
        });
        if (typeof this.props.onChange == 'function') {
            this.props.onChange(this.props.value);
        }
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('value' in nextProps) {
            this.setState({
                value: nextProps.value,
                defaultValue: nextProps.value
            });
        }
    },
    getColorPanel: function getColorPanel() {
        var Panel = this.constructor.Panel;
        var value;
        if (this._isValueRight(_.clone(value))) {
            value = this.state.value;
        } else {
            value = '#fff';
        }
        var pos = this._getPosition();
        return React.createElement(
            'div',
            { className: 'color-panel', style: { top: pos.top } },
            React.createElement(Panel, { defaultColor: value, onChange: this.onChange, ref: 'panel' }),
            React.createElement(
                YYToolbar,
                null,
                React.createElement(
                    YYButton,
                    { type: 'warning', ghost: true, onClick: this._clickReset },
                    '\u6E05\u9664'
                ),
                React.createElement(
                    YYButton,
                    { type: 'primary', ghost: true, onClick: this._clickCancel },
                    '\u53D6\u6D88'
                ),
                React.createElement(
                    YYButton,
                    { type: 'primary', onClick: this._clickOk },
                    '\u786E\u5B9A'
                )
            )
        );
    },
    render: function render() {
        var isShow = this.state.isShow;
        var Panel = isShow ? this.getColorPanel() : null;
        return React.createElement(
            'div',
            { className: this.className() },
            React.createElement('span', { onClick: this.showPanel, className: 'colortrigger', style: { backgroundColor: this.state.value } }),
            Panel
        );
    }
});
module.exports = YYColorPicker;
