// yylib-react/dist/yylib2/base/YYPropTypes.js
"use strict";

/**
 * Created by Administrator on 2016/8/3.
 */
module.exports = {
    array: "array",
    bool: "bool",
    func: "func",
    number: "number",
    object: "object",
    string: "string",
    symbol: "symbol",
    any: "any" //可以是任何格式
    , node: "node" //任何類型的: numbers, strings, elements 或者任何這種類型的陣列
    , element: "element" // React 元素
    , arrayOf: function arrayOf(val) {
        //某种类型的数组
        return { name: 'arrayOf', value: val };
    },
    instanceOf: function instanceOf(val) {
        //某种类别的实体
        return { name: 'instanceOf', value: val };
    },
    objectOf: function objectOf(val) {
        // 具有某種屬性類型的物件(字串類型)
        return { name: 'objectOf', value: val };
    },
    oneOf: function oneOf(val) {
        // 其中一個字串
        return { name: 'oneOf', value: val };
    },
    oneOfType: function oneOfType(val) {
        // 其中一種格式類型
        return { name: 'oneOfType', value: val };
    },
    shape: function shape(val) {
        // 是否符合指定格式的物件
        return { name: 'shape', value: val };
    }
};
