// yylib-react/dist/business/quickdev/api.esstore.js
'use strict';

var React = require('react');
var _ = require('lodash');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass,
    YYMessage = _require.YYMessage,
    YYTable = _require.YYTable;

var ajax = require('yylib-utils/ajax');
var BaseHost = require('../BaseHost');
var BaseWeb = BaseHost + "/icop-elasticsearch-web";
var EsStore = {
    //存储元数据
    saveMetadata: function saveMetadata(options) {
        var index = options.index,
            type = options.type,
            metadataModel = options.metadataModel;

        ajax.postJSON(BaseWeb + '/setMetadata', { index: index, type: type, metadataModel: metadataModel }, success, error);
    },
    //存储数据
    save: function save(options) {
        var index = options.index,
            type = options.type,
            data = options.data,
            success = options.success,
            error = options.error,
            refresh = options.refresh;

        ajax.postJSON(BaseWeb + '/elasticsearch/index', { index: index, type: type, doc: data, refresh: refresh ? refresh : true }, success, error);
    },
    //批量存储数据
    batchSave: function batchSave(options) {
        var index = options.index,
            type = options.type,
            data = options.data,
            success = options.success,
            error = options.error,
            refresh = options.refresh;

        ajax.postJSON(BaseWeb + '/elasticsearch/multiIndex', { index: index, type: type, docList: data, refresh: refresh ? refresh : true }, success, error);
    },
    //更新数据
    update: function update(options) {
        var index = options.index,
            type = options.type,
            id = options.id,
            data = options.data,
            success = options.success,
            error = options.error,
            refresh = options.refresh;

        ajax.postJSON(BaseWeb + '/elasticsearch/update', { index: index, type: type, id: id, doc: data, refresh: refresh ? refresh : true }, success, error);
    },
    //获取单个数据
    getOne: function getOne(options) {
        var index = options.index,
            type = options.type,
            id = options.id,
            success = options.success,
            error = options.error;

        ajax.postJSON(BaseWeb + '/elasticsearch/get', { index: index, type: type, id: id }, success, error);
    },
    //删除单个数据
    deleteOne: function deleteOne(options) {
        var index = options.index,
            type = options.type,
            id = options.id,
            success = options.success,
            error = options.error,
            refresh = options.refresh;

        ajax.delJSON(BaseWeb + '/elasticsearch/delete', { index: index, type: type, id: id, refresh: refresh ? refresh : true }, success, error);
    },
    //删除多个数据
    deleteBatch: function deleteBatch(options) {
        var index = options.index,
            type = options.type,
            ids = options.ids,
            success = options.success,
            error = options.error,
            refresh = options.refresh;

        ajax.delJSON(BaseWeb + '/elasticsearch/multiDelete', { index: index, type: type, ids: ids, refresh: refresh ? refresh : true }, success, error);
    },
    //依据条件查询唯一的单个数据
    findOne: function findOne(options) {
        var index = options.index,
            type = options.type,
            simpleCriteria = options.simpleCriteria,
            orderRules = options.orderRules,
            success = options.success,
            error = options.error;

        ajax.postJSON(BaseWeb + '/elasticsearch/filterForObject', { index: index, type: type, simpleCriteria: simpleCriteria, orderRules: orderRules }, success, error);
    },
    //分页查询数据
    queryList: function queryList(options) {
        var index = options.index,
            type = options.type,
            pageNumber = options.pageNumber,
            pageSize = options.pageSize,
            searchText = options.searchText,
            simpleCriteria = options.simpleCriteria,
            orderRules = options.orderRules,
            success = options.success,
            error = options.error;

        ajax.postJSON(BaseWeb + '/elasticsearch/filterForPage', { index: index, type: type, pageNumber: pageNumber, pageSize: pageSize, searchText: searchText, simpleCriteria: simpleCriteria, orderRules: orderRules }, success, error);
    },
    //查询全部数据
    queryAll: function queryAll(options) {
        var index = options.index,
            type = options.type,
            searchText = options.searchText,
            simpleCriteria = options.simpleCriteria,
            orderRules = options.orderRules,
            success = options.success,
            error = options.error;

        ajax.postJSON(BaseWeb + '/elasticsearch/filterForList', { index: index, type: type, searchText: searchText, simpleCriteria: simpleCriteria, orderRules: orderRules }, success, error);
    }
};
module.exports = EsStore;
