// yylib-react/dist/business/quickdev/DefaultActions.js
'use strict';

var UrlUtils = require('yylib-utils/UrlUtils');
var BaseHost = require('../BaseHost');

var _require = require('./api.all'),
    FormApi = _require.FormApi,
    EditTableApi = _require.EditTableApi;
//系统预置的动作


var DefaultActions = {
    //跳转页面
    gotoPage: function gotoPage(options) {
        var event = options.event,
            action = options.action,
            ui = options.ui;
        var appCode = action.appCode,
            pageCode = action.pageCode;

        var viewUrl = UrlUtils.fillUrlParams('#/pages/view/{appCode}/{pageCode}', {
            appCode: appCode,
            pageCode: pageCode
        });
        window.open(viewUrl, '_self');
    },
    //表单保存
    formSave: function formSave(options) {
        var event = options.event,
            action = options.action,
            ui = options.ui;
        var storeKey = action.storeKey;

        if (storeKey) {
            FormApi.saveData(this, { storeKey: storeKey });
        }
    },
    //表单删除
    formDelete: function formDelete(options) {
        var event = options.event,
            action = options.action,
            ui = options.ui;
        var storeKey = action.storeKey;

        if (storeKey) {
            FormApi.deleteData(this, { storeKey: storeKey });
        }
    },
    //编辑表格-增行
    editTableAddRow: function editTableAddRow(options) {
        var event = options.event,
            action = options.action,
            ui = options.ui;
        var editTableKey = action.editTableKey;

        if (editTableKey) {
            EditTableApi.addRow(this, editTableKey);
        }
    },
    //编辑表格-减行
    editTableDelRow: function editTableDelRow(options) {
        var event = options.event,
            action = options.action,
            ui = options.ui;
        var editTableKey = action.editTableKey;

        if (editTableKey) {
            EditTableApi.delRow(this, editTableKey);
        }
    }
};

module.exports = DefaultActions;
