// yylib-react/dist/business/quickdev/YYRunPage.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * Created by Dio on 2017/9/12.
 */
var React = require('react');
var _ = require('lodash');
var YYCreatePage = require('./YYCreatePage');
var ui = require('yylib-ui');

var YY = {
    modules: {},
    pages: {},
    ui: ui
};
YY.define = function (moduleName, moduleFunc) {
    YY.modules[moduleName] = moduleFunc;
};
YY.require = function (moduleName) {
    return YY.modules[moduleName];
};

YY.createPage = function (pageCode, options) {
    YY.pages[pageCode] = function () {
        var _options = options;
        return _options;
    };
};

YY.getPage = function (pageCode) {
    return YY.pages[pageCode];
};

var loadConfig = {
    "moduleA": "./ModuleA.js",
    "moduleB": "./ModuleB.js",
    "P000111": "./A.js",
    "P000112": "./B.js"
};

for (var loadKey in loadConfig) {};

(function (yy) {
    //console.log('运行-----js---a');
    //var a = "aaa";
    //var {YYButton} = yy.ui;
    //console.log('YYButton---->',YYButton);
    //yy.createPage("P000111",{
    //    // 添加行
    //    addRowBtn:{
    //        "onClick":function () {
    //            var edi = this.findUI('editTable');
    //            edi.api.addRow();
    //            console.log('--addRow-'+a);
    //        }
    //    },
    //    // 删除行
    //    removeRowBtn:{
    //        "onClick":function () {
    //            var edi = this.findUI('editTable');
    //            var selectedRowKeys = edi.api.getSelectedRowKeys();
    //            edi.api.delRow(selectedRowKeys);
    //            console.log('--delRow-');
    //        }
    //    }
    //});
})(YY);

(function (yy) {
    //console.log('运行-----js---B');
    //var BB ="BBB";
    //var b = function(){
    //    console.log('执行B----》'+BB);
    //}
    //yy.createPage("P000112",{
    //    // 添加行
    //    addRowBtn:{
    //        "onClick":function () {
    //            var edi = this.findUI('editTable');
    //            edi.api.addRow();
    //            b();
    //            console.log('--addRow2-');
    //        }
    //    },
    //    // 删除行
    //    removeRowBtn:{
    //        "onClick":function () {
    //            var edi = this.findUI('editTable');
    //            var selectedRowKeys = edi.api.getSelectedRowKeys();
    //            edi.api.delRow(selectedRowKeys);
    //            console.log('--delRow2-');
    //        }
    //    }
    //});
})(YY);
//router进入页面前的回调
function initRunPageRouter(nextState, replace) {}

var YYRunPage = React.createClass({
    displayName: 'YYRunPage',

    statics: {
        enterPage: initRunPageRouter
    },
    getDefaultProps: function getDefaultProps() {
        return {};
    },
    getInitialState: function getInitialState() {
        return {};
    }
    //,componentWillReceiveProps(nextProps){
    //}
    //,componentDidMount: function () {
    //}
    , render: function render() {
        var _props2 = this.props,
            router = _props2.router,
            location = _props2.location,
            params = _props2.params,
            route = _props2.route,
            routeParams = _props2.routeParams,
            routes = _props2.routes;

        var _props = { router: router, location: location, params: params, route: route, routeParams: routeParams, routes: routes };
        var myPageEvent = YY.getPage('P000112');
        myPageEvent = myPageEvent || function () {};
        //console.log('myPageEvent--->',myPageEvent);
        return React.createElement(YYCreatePage, _extends({}, _props, { uiEvent: myPageEvent() }));
    }
});
module.exports = YYRunPage;
