// yylib-react/dist/yylib2/refer/YYReferOperateDlg.js
'use strict';

var React = require('react');
var YYClass = require('../base/YYClass');
var YYModal = require('../modal/YYModal');
var field = YYClass.field;
var PropTypes = YYClass.PropTypes;
//定义META
var uiMeta = {
	name: 'refer-foot',
	fields: [
	//field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
	field('operateurl', null, PropTypes.string, true, 'operateurl'), field('show', false, PropTypes.string, true, '是否显示')]
};
var YYReferOperateDlg = YYClass.create({
	render: function render() {
		var that = this;
		var docHeight = document.documentElement.clientHeight;
		var dlgWidth = document.documentElement.clientWidth * 0.90;
		var dlgHeight = document.documentElement.clientHeight * 0.85;
		return React.createElement(
			YYModal,
			{
				width: dlgWidth,
				style: { height: dlgHeight, top: docHeight * 0.05 },
				footer: null,
				closable: true,
				onCancel: that.props.onCancel,
				maskClosable: true,
				className: this.className('modal'),
				visible: this.props.show },
			React.createElement('iframe', { src: that.props.operateurl, style: { width: dlgWidth, height: dlgHeight, border: 'none' } })
		);
	}
});

module.exports = YYReferOperateDlg;
