// yylib-react/dist/yylib2/refer/YYReferTreeGrid.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _YYTreeInRefer = require('./YYTreeInRefer');

var _YYTreeInRefer2 = _interopRequireDefault(_YYTreeInRefer);

var _YYReferOperateDlg = require('./YYReferOperateDlg');

var _YYReferOperateDlg2 = _interopRequireDefault(_YYReferOperateDlg);

var _reactRedux = require('react-redux');

var _reference = require('./actions/reference');

var _ReferConstants = require('./constants/ReferConstants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
** 树表型参照：由YYReferDialog调用
**/
var React = require('react');
var _ = require('lodash');
var YYForm = require('../formcontrols/form/YYForm');
var YYFormItem = require('../formcontrols/form/YYFormItem');
var ajax = require('../../utils/ajax');
var YYClass = require('../base/YYClass');
var Tabs = require('../tabs/YYTabs');
var TabPane = require('../tabs/YYTab');
var Tag = require('../tag/YYTag');
var Icon = require('../icon/YYIcon');
var YYButton = require('../button/YYButton');
var YYTable = require('../table/YYTable');
var Pagination = require('../paging/YYPagination');
var YYScrollbar = require('../scrollbar/YYScrollbar');
var Select = require('../formcontrols/select/YYSelect');
var YYFootInRefer = require('./YYFootInRefer');
require('./style/index.less');
var YYRow = require('../layout/YYRow');
var YYCol = require('../layout/YYCol');
var YYInputButton = require('../formcontrols/inputbutton/YYInputButton');
var YYMessage = require('../message/YYMessage');
var AuthToken = require("../../utils/AuthToken");
var TreeUtils = require("../../utils/TreeUtils");

var PropTypes = YYClass.PropTypes;
var field = YYClass.field;

var _require = require('./utils/YYReferUtils'),
    getColumns = _require.getColumns,
    ifInArray = _require.ifInArray,
    removeRow = _require.removeRow,
    getPagination = _require.getPagination,
    getDefaultRows = _require.getDefaultRows,
    getDefaultKeys = _require.getDefaultKeys,
    getHeaders = _require.getHeaders,
    formatOftenToServer = _require.formatOftenToServer,
    formatOftenFromServer = _require.formatOftenFromServer,
    getOftenKeys = _require.getOftenKeys,
    sortWithTableRows = _require.sortWithTableRows;
// const pageSize = 5;

var store = window.localStorage;
var tableKey = null;
var treeNodeKey = null; //key：store中存储的之前select的树节点
var selectedTreeKey = null; //值：选择的树节点id。初始值从store中获取， 根据treeNodeKey key值获取
var treeNode = null; //key: store中存储的之前的select的树节点信息
var selectedTreeNode = null; //值： 选择的树节点id。初始值从store中获取， 根据treeNode key值获取
var storeKeyOfTree = null; // key: treeURL
var charCode = String.fromCharCode(2);
var relyfield = null;
var ifSelectDefault = true;

function noop() {};

var mapStateToProps = function mapStateToProps(state) {
	return {
		refinfo: state.refinfo,
		loadingReferTree: state.refinfo.loadingReferTree,
		loadingReferGrid: state.refinfo.loadingReferGrid
	};
};
//定义META
var uiMeta = {
	name: 'refer-treegrid',
	fields: [
	//field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
	field('refinfokey', null, PropTypes.string, true, '参照key'), field('refname', null, PropTypes.string, true, '参照名称'), field('dataurl', null, PropTypes.string, true, '查询url'), field('operateurl', null, PropTypes.string, true, '操作url'), field('treerelyurl', null, PropTypes.string, true, '树参照url'), field('gridrelyurl', null, PropTypes.string, true, '表参照url'), field('initialValue', null, PropTypes.oneOfType([PropTypes.object, PropTypes.Array]), false, '默认初始值'), field('relyfield', null, PropTypes.array, true, '参照字段'), field('orderfield', null, PropTypes.string, true, '排序字段'), field('often', null, PropTypes.bool, true, '是否显示常用'), field('multiselect', null, PropTypes.bool, true, '是否支持多选'), field('querystr', null, PropTypes.string, true, '查询条件(过期)'), field('condition', null, PropTypes.oneOfType([PropTypes.func, PropTypes.object]), true, '自定义查询条件'), field('onOk', null, PropTypes.func, true, '确认操作回调函数'), field('onCancel', null, PropTypes.func, true, '取消操作回调函数'), field('customheader', null, PropTypes.element, false, '自定义header'), field('contentH', null, PropTypes.number, true, '内容区域的高度'), field('gridheaders', null, PropTypes.array, false, '表头信息'), field('treeValue', null, PropTypes.string, false, '树表参照中，树默认初始值节点id'), field('multiTreeValue', null, PropTypes.Array, false, '树表参照(树切换场景)中，树默认初始值节点ids'), field('tagNodes', null, PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]), false, '默认高亮绿色显示的节点'), field('searchPlaceholder', null, PropTypes.string, false, '搜索框显示信息'), field('lazyTree', false, PropTypes.bool, false, '是否启用异步树功能'), field('rowClickCheckable', true, PropTypes.bool, false, 'checkable启用时，点击行是否同时触发点击勾选框')]
};
var YYReferTreeGrid = YYClass.create({
	tableData: null,
	uiMeta: uiMeta,
	getInitialState: function getInitialState() {
		return {
			current: 1,
			pageSize: 10,
			refid: null,
			name: null,
			loadOften: false,
			selectedNode: null,
			selectedNodeKey: null,
			selectedRow: this.props.multiselect ? null : getDefaultRows(this.props.initialValue, this.props.idField, this.props.codeField, this.props.nameField, this.props.multiselect), //单选row
			selectedRowKey: this.props.multiselect ? null : getDefaultKeys(this.props.initialValue, this.props.multiselect), //单选key,多选模式下不获取
			selectedRows: this.props.multiselect ? getDefaultRows(this.props.initialValue, this.props.idField, this.props.codeField, this.props.nameField, this.props.multiselect) : null, //多选rows
			selectedRowKeys: this.props.multiselect ? getDefaultKeys(this.props.initialValue, this.props.multiselect) : null, //多选keys, 单选模式下不获取
			searchVal: null,
			showOperateDialog: false, //是否打开operate dialog
			operateurl: null //operate的url
		};
	},
	componentWillMount: function componentWillMount() {
		var _props = this.props,
		    refinfokey = _props.refinfokey,
		    refinfo = _props.refinfo,
		    pageSize = _props.pageSize;

		if (store) {
			var userCode = AuthToken.getUserCode();
			//常用数据key
			tableKey = _ReferConstants.REFER_TABLEKEY + userCode + "-" + refinfokey;
			//优先从pageSize获取
			var pageSizeBefore = YYTable.getLocalPageSize(tableKey);
			// 当打开参照，并且本地无pagesize记录时，以参照注册信息中的pagesize为准，
			// 或者当pageSize发生改变时
			if (!(pageSizeBefore && pageSizeBefore > 0) || parseInt(pageSizeBefore) !== pageSize) {
				//参照注册信息未设置pagesize时，默认10
				var pageSizeProps = this.props.pageSize ? this.props.pageSize : 10;
				YYTable.setLocalPageSize(tableKey, pageSizeProps);
			}

			treeNodeKey = _ReferConstants.REFER_TREEGRID_SELECTED_NODEKEY + userCode + '-' + refinfokey;
			selectedTreeKey = store[treeNodeKey] ? store[treeNodeKey] : null;
			treeNode = _ReferConstants.REFER_TREEGRID_SELECTED_NODE + userCode + '-' + refinfokey;
			selectedTreeNode = store[treeNode] ? JSON.parse(store[treeNode]) : null;
			//树选择urlkey
			storeKeyOfTree = _ReferConstants.REFER_SELECTED_TREEURL + userCode + "-" + refinfokey;
		}
	},

	// 根据当前选择的tree，返回对应的relyfield，以查询表数据时用
	_getRelyField: function _getRelyField(treerelyselect, currentTree) {
		var relyfield = this.props.relyfield;
		for (var i = 0; i < treerelyselect.length; i++) {
			if (treerelyselect[i].treerelyurl === currentTree) {
				return treerelyselect[i].relyfield;
			}
		}
		// 避免控制，当字表中没有注册relyfield，仍然采用主表中的relyfield
		return relyfield;
	},
	componentDidMount: function componentDidMount() {
		var _props2 = this.props,
		    treerelyurl = _props2.treerelyurl,
		    treerelyselect = _props2.treerelyselect;

		var firstTree = treerelyurl;
		relyfield = this.props.relyfield;
		if (treerelyselect && treerelyselect instanceof Array && treerelyselect.length > 0) {
			firstTree = store[storeKeyOfTree] ? JSON.parse(store[storeKeyOfTree]) : treerelyselect[0].treerelyurl;
			// 树切换情况下， relyfield采用字表中的
			relyfield = this._getRelyField(treerelyselect, firstTree);
		}
		this._doQueryTree(firstTree);
	},

	// 根据默认的树url得出tree的id
	_getTreeId: function _getTreeId(url, trees) {
		for (var i = 0; i < trees.length; i++) {
			if (trees[i]['treerelyurl'] === url) {
				return trees[i]['id'];
			}
		}
		return trees[0]['id'];
	},

	// 根据treeid， 从默认值中获取对应的默认节点值
	_getSelectedTreeKey: function _getSelectedTreeKey(treeId, treeValues) {
		for (var i = 0; i < treeValues.length; i++) {
			if (treeValues[i]['treeId'] === treeId) {
				return treeValues[i]['nodeValue'];
			}
		}
		return null;
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		if (nextProps && nextProps.refinfo) {
			var refinfo = nextProps.refinfo;

			if (nextProps.refinfo.doReload) {
				//重新查询数据  与搜索的功能保持一致
				var _props3 = this.props,
				    dataurl = _props3.dataurl,
				    querystr = _props3.querystr,
				    condition = _props3.condition,
				    orderfield = _props3.orderfield,
				    refid = _props3.refid,
				    dispatch = _props3.dispatch;

				var query = this.getQuery(querystr, condition, relyfield, null, orderfield, 1);
				dispatch((0, _reference.getGridData)(dataurl, query, _ReferConstants.REFER_TYPE_TREEGRID));
				//重置搜索条件和页码
				this.refs.searchGrid.state.value = '';
				this.setState({
					current: 1
				});
			}
			if (refinfo.showMsg) {
				var showMsg = refinfo.showMsg;
				if (showMsg.success) {
					YYMessage.success(showMsg.success);
				} else if (showMsg.error) {
					YYMessage.error(showMsg.error);
				}
			}
			if (ifSelectDefault) {
				var nodes = refinfo.tree;
				var newSelectedTreeKey = selectedTreeKey;
				//add by zhangzhzhc 2017/05/15 start 对默认值进行验证，是否能找到对应的树节点，以应对传入的查询左树条件变化的情形
				if (nodes && newSelectedTreeKey !== null && newSelectedTreeKey !== '' && newSelectedTreeKey !== 'null') {
					selectedTreeNode = TreeUtils.findById(nodes, newSelectedTreeKey);
					if (!selectedTreeNode) {
						newSelectedTreeKey = null;
					}
				}
				//add by zhangzhzhc 2017/05/15 end

				// 初始化后，默认选中之前选中的值。否则，默认选中用户传递的初始值，再否，默认选择第一个子节点，没有子的情况默认选中根节点
				if (newSelectedTreeKey === null || newSelectedTreeKey === '' || newSelectedTreeKey === 'null') {
					var multiTreeValue = nextProps.multiTreeValue;
					var treerelyselect = nextProps.treerelyselect;
					if (multiTreeValue && multiTreeValue instanceof Array && multiTreeValue.length > 0 && treerelyselect && treerelyselect instanceof Array && treerelyselect.length > 0) {
						//树切换场景
						//多树场景，取当前树
						var choosedTreeId = store[storeKeyOfTree] ? this._getTreeId(JSON.parse(store[storeKeyOfTree]), treerelyselect) : treerelyselect[0]['id'];
						//取树节点 如果有用户配置的默认值，取默认值
						newSelectedTreeKey = this._getSelectedTreeKey(choosedTreeId, multiTreeValue);
						selectedTreeNode = TreeUtils.findById(nodes, newSelectedTreeKey);
						// selectedTreeKey = selectedTreeNode?selectedTreeNode['nodeValue']:null;
					}
					if (newSelectedTreeKey === null || newSelectedTreeKey === '' || newSelectedTreeKey === 'null') {
						if (this.props.treeValue) {
							// 普通树表场景 如果有默认值取默认值
							newSelectedTreeKey = this.props.treeValue;
							// todo 算法实现 遍历树
							selectedTreeNode = TreeUtils.findById(nodes, newSelectedTreeKey);
						} else if (refinfo) {
							// 无树默认值场景
							if (nodes && nodes[0]) {
								if (nodes[0].children && nodes[0].children[0]) {
									// 选择第一个根节点的第一个子节点
									newSelectedTreeKey = nodes[0].children[0].id;
									selectedTreeNode = nodes[0].children[0];
								} else {
									// 当没有子节点时，默认选中根节点(如：切换组织机构的场景)
									newSelectedTreeKey = nodes[0].id;
									selectedTreeNode = nodes[0];
								}
							}
						}
					}
				}
				// 补充nodes在初始几种状态下为null时， selectedTreeNode未赋值的处理
				if (nodes !== null && newSelectedTreeKey !== null && selectedTreeNode === null) {
					selectedTreeNode = TreeUtils.findById(nodes, newSelectedTreeKey);
				}
				if (selectedTreeNode !== null && selectedTreeNode.value === undefined) {
					// 直接从树数据中取的情况， name-->value
					selectedTreeNode.value = selectedTreeNode.name;
				}
				if (newSelectedTreeKey !== selectedTreeKey) {
					selectedTreeKey = newSelectedTreeKey;
					if (this.props.onNodeSelect && typeof this.props.onNodeSelect === 'function') {
						this.props.onNodeSelect(selectedTreeNode);
					}
				}
				this.setState({
					selectedNodeKey: selectedTreeKey ? selectedTreeKey : refinfo.treegrid
				});
			}
		}
	},
	componentWillUnmount: function componentWillUnmount() {
		// 存储，以下次打开时默认选中
		store.setItem(treeNodeKey, selectedTreeKey);
		store.setItem(treeNode, JSON.stringify(selectedTreeNode));
	},

	//组件内公共方法，执行查询树数据
	_doQueryTree: function _doQueryTree(treeUrl) {
		var _props4 = this.props,
		    dataurl = _props4.dataurl,
		    treerelyurl = _props4.treerelyurl,
		    treerelyselect = _props4.treerelyselect,
		    querystr = _props4.querystr,
		    condition = _props4.condition,
		    orderfield = _props4.orderfield,
		    dispatch = _props4.dispatch;

		var query = this.getQuery(querystr, condition, null, null, orderfield, 1);
		var queryGrid = this.getQuery(querystr, condition, relyfield, null, orderfield, 1, null, false);
		ifSelectDefault = true;
		// 延迟，解决tab错位的bug
		dispatch((0, _reference.loadingReferTree)(true));
		dispatch((0, _reference.loadingReferGrid)(true));
		setTimeout(function () {
			// 查询树后根据默认选择的树节点查询grid数据
			dispatch((0, _reference.getTreeData)(treeUrl, query, _ReferConstants.REFER_TYPE_TREEGRID, { 'dataurl': dataurl, 'querydata': queryGrid, 'relyfield': relyfield, 'nodeKey': selectedTreeKey }));
		}, 500);
	},
	changeTags: function changeTags(key) {
		this.setState({
			loadOften: !this.state.loadOften
		});
	},

	// 保存/删除 常用数据到服务器. oper=0: 手动， oper=1：自动
	_saveOften: function _saveOften(records, oper, isDelete) {
		var dispatch = this.props.dispatch;
		var userId = AuthToken.getUserId();
		var refinfokey = this.props.refinfokey;
		var idField = this.props.idField;
		var serverUrl = this.props.serverUrl;
		if (isDelete === true) {
			var ids = records[idField]; // 当前删除只支持单个删除，后续支持批量删除后在此扩展
			dispatch((0, _reference.deleteOften)(serverUrl, { userId: userId, refinfokey: refinfokey, ids: ids }));
		} else {
			var dataset = formatOftenToServer(records, idField);
			if (dataset && dataset.length > 0) {
				dispatch((0, _reference.addOften)(serverUrl, { userId: userId, refinfokey: refinfokey, oper: oper, dataset: dataset }));
			}
		}
	},

	//添加常用
	addOften: function addOften(record, e) {
		e.stopPropagation();
		if (store && record) {
			var nodeKey = record[this.props.idField];
			this._saveOften(record, 0);
		}
		this.setState({
			loadOften: !this.state.loadOften
		});
	},

	//删除常用
	delOften: function delOften(record, e) {
		e.stopPropagation();
		var nodeKey = record[this.props.idField];
		var isDelete = true; // 便于理解和维护。此处用于标识删除操作
		this._saveOften(record, null, isDelete);

		this.setState({
			loadOften: !this.state.loadOften
		});
	},

	//确认
	btn_ok: function btn_ok(e) {
		this.doHandleDatas(null, null, true);
	},

	//取消
	btn_cancel: function btn_cancel(e) {
		if (this.props.onCancel) {
			this.props.onCancel(e);
		}
	},
	onClean: function onClean() {
		this.setState({
			selectedRow: null,
			selectedRowKey: null,
			selectedRows: null,
			selectedRowKeys: null

		});
	},

	// 处理选择的树节点变化后的逻辑
	_changeSelectNode: function _changeSelectNode(node) {
		ifSelectDefault = false;
		this.setState({
			current: 1,
			selectedNodeKey: selectedTreeKey,
			selectedNode: node,
			selectedRow: null,
			selectedRowKey: null
		});
		var _props5 = this.props,
		    dataurl = _props5.dataurl,
		    querystr = _props5.querystr,
		    condition = _props5.condition,
		    orderfield = _props5.orderfield,
		    refid = _props5.refid,
		    dispatch = _props5.dispatch;

		var query = this.getQuery(querystr, condition, relyfield, null, orderfield, 1);
		dispatch((0, _reference.getGridData)(dataurl, query, _ReferConstants.REFER_TYPE_TREEGRID));
		if (this.props.onNodeSelect && typeof this.props.onNodeSelect === 'function') {
			this.props.onNodeSelect(node);
		}
	},

	//点击树节点
	onSelectNode: function onSelectNode(info, e) {
		var treeId = null;
		if (info.length > 0) {
			treeId = info[0];
			selectedTreeKey = treeId;
			var curNode = {
				id: e.selectedNodes[0].key,
				code: e.selectedNodes[0].props.code,
				value: e.selectedNodes[0].props.title,
				extdata: e.selectedNodes[0].props.extdata
			};
			selectedTreeNode = curNode;
			this._changeSelectNode(curNode);
		} else {
			// 取消选中之前选中后的节点后， 默认勾选第一个子节点或者根节点
			var curNode = {};
			var nodes = this.props.refinfo.tree;
			if (nodes && nodes[0]) {
				if (nodes[0].children && nodes[0].children[0]) {
					// 选择第一个根节点的第一个子节点
					selectedTreeKey = nodes[0].children[0].id;
					curNode = nodes[0].children[0];
				} else {
					// 当没有子节点时，默认选中根节点(如：切换组织机构的场景)
					selectedTreeKey = nodes[0].id;
					curNode = nodes[0];
				}
				this._changeSelectNode(curNode);
			}
		}
	},

	//设置查询条件
	// 查询条件(旧)、查询条件、树节点字段名、搜索值、排序、页码、每页条数、是否拼接当前所选树节点为条件
	getQuery: function getQuery(querystr, condition, relyfield, searchText, orderfield, pageNumber, pageSize, ifConcatRelyId) {
		var query = null;
		var relystr = null;
		var relyid = selectedTreeKey;
		if (relyid && relyfield) relystr = " " + relyfield + "=" + relyid;
		if (querystr) query = querystr;
		if (ifConcatRelyId !== false) {
			if (query) {
				if (relystr) query += " and " + relystr;
			} else {
				query = relystr;
			}
		}
		var queryObj = {};
		if (pageNumber && pageSize !== -1) {
			queryObj.pageSize = pageSize ? pageSize : YYTable.getDefaultPageSize(tableKey);
			queryObj.pageNumber = pageNumber;
		}
		if (query && query.length > 0) queryObj.relyCondition = query;
		if (orderfield && orderfield.length > 0) queryObj.orderCondition = orderfield;
		if (searchText && searchText.length > 0) queryObj.searchText = searchText;
		var _condition = ajax.getParams(condition);
		if (!_.isEmpty(_condition) && _.isPlainObject(_condition)) {
			//转成JSON格式字符串
			queryObj.condition = JSON.stringify(_condition);
		}
		return queryObj;
	},

	//处理树数据的子数据
	getChildren: function getChildren(childrenData) {
		var _this = this;

		var data = [];
		if (childrenData && childrenData.length > 0) {
			childrenData.map(function (elem, eindex) {
				var node = _.cloneDeep(elem);
				node.key = elem.id;
				if (elem.children && elem.children.length > 0) {
					node.children = _this.getChildren(elem.children);
				} else {
					node.isLeaf = node.isLeaf == null ? true : node.isLeaf; //如果isLeaf存在，就保留原始值， 否则设置isLeaf=true;
				}
				data.push(node);
			});
		}
		return data;
	},

	//页码更改
	onChangePage: function onChangePage(current) {
		var value = this.refs.searchGrid.state.value;
		var _props6 = this.props,
		    dataurl = _props6.dataurl,
		    querystr = _props6.querystr,
		    condition = _props6.condition,
		    orderfield = _props6.orderfield,
		    refid = _props6.refid,
		    dispatch = _props6.dispatch;

		var query = this.getQuery(querystr, condition, relyfield, value, orderfield, current);
		dispatch((0, _reference.getGridData)(dataurl, query, _ReferConstants.REFER_TYPE_TREEGRID));
		this.setState({
			current: current
		});
	},

	// 每页显示条数更改
	onChangePageSize: function onChangePageSize(current, pageSize) {
		var value = this.refs.searchGrid.state.value;
		var _props7 = this.props,
		    dataurl = _props7.dataurl,
		    querystr = _props7.querystr,
		    condition = _props7.condition,
		    orderfield = _props7.orderfield,
		    refid = _props7.refid,
		    dispatch = _props7.dispatch;

		var query = this.getQuery(querystr, condition, relyfield, value, orderfield, current, pageSize);
		dispatch((0, _reference.getGridData)(dataurl, query, _ReferConstants.REFER_TYPE_TREEGRID));
		this.setState({
			current: current
		});
	},

	//搜索表
	onSearchGrid: function onSearchGrid(e) {
		var value = this.refs.searchGrid.state.value;
		if (this.props.loadingReferGrid) {
			return;
		} else {
			var _props8 = this.props,
			    dataurl = _props8.dataurl,
			    querystr = _props8.querystr,
			    orderfield = _props8.orderfield,
			    condition = _props8.condition,
			    refid = _props8.refid,
			    dispatch = _props8.dispatch;

			var query = this.getQuery(querystr, condition, relyfield, value, orderfield, 1);
			dispatch((0, _reference.getGridData)(dataurl, query, _ReferConstants.REFER_TYPE_TREEGRID));
			this.setState({
				current: 1,
				searchVal: value
			});
		}
	},

	//公用方法
	doHandleDatas: function doHandleDatas(selectedRows, selectedRow, isHandleOk) {
		var multiselect = this.props.multiselect;

		var row = null;
		var curTreeNode = selectedTreeNode;
		if (multiselect) {
			row = selectedRows ? selectedRows : this.state.selectedRows;
		} else {
			// curTreeNode = selectedTreeNode;
			row = selectedRow ? selectedRow : this.state.selectedRow;
		}
		if (isHandleOk) {
			if (this.props.often) {
				this._saveOften(row, 1);
			}
			this.props.onOk(row, curTreeNode);
		} else {
			this.props.onSelectedChanged(row);
		}
	},

	//获取所选数据API：调用者根据this.refs.设置的ref名.getCurrentSelected来获取数据
	getCurrentSelected: function getCurrentSelected() {
		var state = this.state;
		if (this.props.multiselect) {
			return {
				fullValue: state.selectedRows
			};
		} else {
			return {
				fullValue: state.selectedRow,
				curTreeNode: selectedTreeNode
			};
		}
	},
	onChangeRow: function onChangeRow(keys, rows) {
		// console.log('rows', rows)
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys ? this.state.selectedRowKeys : [];
		if (rows) {
			var sortedRows = sortWithTableRows(rows, this.tableData.props.dataSource, this.tableData.props.pagination.current, this.props.idField);
			for (var i = 0; i < sortedRows.length; i++) {
				if (!ifInArray(selectedRowKeys, sortedRows[i][this.props.idField])) {
					//确定当前元素插入的位置index
					var index = 0;
					if (sortedRows.indexOf(sortedRows[i]) == 0) {
						index = 0;
					} else {
						index = selectedRowKeys.indexOf(selectedRowKeys[i - 1]) + 1;
					}
					//根据页数推算出当前元素的位置indexByPage
					var indexByPage = 0;
					if (sortedRows[i].pageCurrent == 1) {
						indexByPage = 0;
					} else {
						for (var t = 0; t < selectedRows.length; t++) {
							// console.log(sortedRows[t].pageCurrent, selectedRows[t].pageCurrent)
							if (sortedRows[i].pageCurrent > selectedRows[t].pageCurrent) indexByPage++;else break;
						}
					}
					//元素的具体位置是index+indexByPage
					selectedRowKeys.splice(index + indexByPage, 0, sortedRows[i][this.props.idField]);
					selectedRows.splice(index + indexByPage, 0, sortedRows[i]);
				};
			}
		} else {
			selectedRows = null;
		}
		this.setState({
			selectedRows: selectedRows,
			selectedRowKeys: selectedRowKeys
		});
		this.doHandleDatas(selectedRows, null, false);
	},

	//勾选
	onSelectRow: function onSelectRow(record, selected, selectRows) {
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys;
		if (!selected) {
			var key = record[this.props.idField];
			selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
			selectedRows = removeRow(selectedRows, key, this.props.idField);
			this.setState({
				selectedRows: selectedRows,
				selectedRowKeys: selectedRowKeys
			});
			this.doHandleDatas(selectedRows, null, false);
		}
	},

	//全勾选
	onSelectAllRow: function onSelectAllRow(selected, selectRows, changeRows) {
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys;
		if (!selected) {
			for (var i = changeRows.length - 1; i >= 0; i--) {
				var key = changeRows[i][this.props.idField];
				selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
				selectedRows = removeRow(selectedRows, key, this.props.idField);
			};
			this.setState({
				selectedRows: selectedRows,
				selectedRowKeys: selectedRowKeys
			});
			this.doHandleDatas(selectedRows, null, false);
		}
	},

	//选中行 单选时
	onRowClick: function onRowClick(record, index) {
		this.setState({
			selectedRow: record,
			selectedRowKey: record[this.props.idField]
		});
		this.doHandleDatas(null, record, false);
	},

	// 选中行 双击时 单选时
	onRowDoubleClick: function onRowDoubleClick(record, index) {
		this.doHandleDatas(null, record, true);
	},

	//关闭标签
	onTagClose: function onTagClose(key) {
		var selectedRows = this.state.selectedRows;
		var selectedRowKeys = this.state.selectedRowKeys;
		selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
		removeRow(selectedRows, key, this.props.idField);
		this.setState({
			selectedRows: selectedRows,
			selectedRowKeys: selectedRowKeys
		});
		this.doHandleDatas(selectedRows, null, false);
	},

	//点击带链接字段
	onClickOperate: function onClickOperate(record, e) {
		if (record) {
			var url = null;
			var operateurl = this.props.operateurl;
			if (operateurl) {
				if (operateurl.indexOf('{') > 0 && operateurl.indexOf('}') > 0) {
					// 当注册的url有占位符{}，则按照注册的参数组织url, relyfield、selectedNodeId是约定名称
					record[_ReferConstants.REFER_PARAM_RELYFIELD] = relyfield; // relyfield
					record[_ReferConstants.REFER_PARAM_SELECTEDNODEID] = selectedTreeKey; // 选择的树节点id
					url = ajax.fillUrlParams(operateurl, record);
				} else {
					// 当注册的url没有占位符，默认参数是当前数据的id
					url = operateurl + record[this.props.idField];
				}
				this.setState({
					showOperateDialog: true,
					operateurl: url
				});
			}
		}
	},

	//关闭点击链接打开的dialog回调
	onCloseOperate: function onCloseOperate() {
		this.setState({
			showOperateDialog: false
		});
	},

	// 切换树 树组件选择下拉框onchange回调
	onTreeChanged: function onTreeChanged(newTree) {
		relyfield = this._getRelyField(this.props.treerelyselect, newTree);
		ifSelectDefault = true; // 切换后默认选中第一个树节点
		// TODO 此处如果有对应默认值，则切换到对应默认值，否则清空, 以切换后默认选中第一个根节点
		selectedTreeKey = null;
		store.setItem(storeKeyOfTree, JSON.stringify(newTree));
		this._doQueryTree(newTree);
	}
	//异步加载树节点
	,
	_loadData: function _loadData(treeNode) {
		var _this2 = this;

		var _props9 = this.props,
		    querystr = _props9.querystr,
		    condition = _props9.condition,
		    orderfield = _props9.orderfield,
		    treerelyurl = _props9.treerelyurl,
		    dispatch = _props9.dispatch;

		var queryData = this.getQuery(querystr, condition, null, null, orderfield, 1);
		return new Promise(function (resolve) {
			(0, _reference.getLazyTreeData)({
				dispatch: dispatch,
				dataurl: treerelyurl,
				queryData: queryData,
				treeData: _this2.props.refinfo.tree,
				curKey: treeNode.props.eventKey,
				resolve: resolve
			});
		});
	},
	//渲染tabs内容
	_renderTabs: function _renderTabs() {
		var _this3 = this;

		var that = this;
		var _props10 = this.props,
		    refinfo = _props10.refinfo,
		    refname = _props10.refname,
		    multiselect = _props10.multiselect;

		var grid = refinfo.treegrid;
		var header = getHeaders(that.props.gridheaders, grid);

		var nameField = this.props.nameField;
		var idField = this.props.idField;

		var griddata = grid ? grid.data : null;
		var totalsize = griddata ? griddata.count : 0; //总数量
		var content = griddata ? griddata.content : null;
		var data = [];
		if (content && content.length > 0) {
			data = content;
			data.map(function (elem, eindex) {
				elem.key = elem[_this3.props.idField];
			});
		}

		var treeData = [];
		var nodes = refinfo.tree;
		if (nodes && nodes.length > 0) {
			nodes.map(function (elem, eindex) {
				var node = _.cloneDeep(elem);
				node.key = elem.id;
				node.children = _this3.getChildren(elem.children);
				treeData.push(node);
			});
		}

		var columnsInStore = getColumns(header, store, this.props.nameField, this.props.operatefield, this.props.operateurl, this.onClickOperate);
		var columns = columnsInStore.columns;
		var oftenColumns = columnsInStore.oftenColumns;
		var totalWidth = columnsInStore.totalWidth;

		var selectedNode = this.state.selectedNode;
		var rowSelection = null;
		if (multiselect) {
			totalWidth += _ReferConstants.REFER_CHECKBOX_WIDTH;
			rowSelection = {
				onChange: this.onChangeRow,
				onSelect: this.onSelectRow,
				onSelectAll: this.onSelectAllRow,
				selectedRowKeys: this.state.selectedRowKeys
			};
		}

		var oftenTab = null;
		if (this.props.often) {
			//读取常用数据
			var oftenData = formatOftenFromServer(this.props.refinfo.often);
			var oftenkeys = getOftenKeys(this.props.refinfo.often);
			columns.push({
				key: 'operation',
				title: _ReferConstants.REFER_ADD_OFTEN,
				width: _ReferConstants.REFER_OPERATE_COLUMN_WIDTH,
				render: function render(text, record) {
					return React.createElement(
						'div',
						null,
						ifInArray(oftenkeys, record[idField]) ? React.createElement(YYButton, { icon: 'minus', ghost: true, onClick: _this3.delOften.bind(null, record) }) : React.createElement(YYButton, { icon: 'plus', ghost: true, onClick: _this3.addOften.bind(null, record) })
					);
				}
			});
			oftenColumns.push({
				key: 'operationOften',
				title: _ReferConstants.REFER_DELETE_OFTEN,
				width: _ReferConstants.REFER_OPERATE_COLUMN_WIDTH,
				render: function render(text, record) {
					return React.createElement(YYButton, { icon: 'minus', ghost: true, onClick: _this3.delOften.bind(null, record) });
				}
			});

			totalWidth += _ReferConstants.REFER_OPERATE_COLUMN_WIDTH;
			oftenTab = React.createElement(
				TabPane,
				{ key: '1', tab: '\u5E38\u7528', disabled: !this.props.often, className: 'yyui-refer-tab' },
				React.createElement(YYTable, { columns: oftenColumns,
					style: { marginTop: 5 },
					scroll: { x: totalWidth, y: this.props.contentH },
					rowKey: that.props.idField,
					dataSource: oftenData, bordered: true,
					pagination: false,
					rowSelection: rowSelection,
					rowActiveKey: that.state.selectedRowKey,
					onRowDoubleClick: multiselect ? noop : this.onRowDoubleClick,
					onRowClick: multiselect ? noop : this.onRowClick,
					rowClickCheckable: this.props.rowClickCheckable })
			);
		}
		var gridTableH = this.props.contentH - 28 - 45 - 10 - 10; //28是表格上方搜索框的高度、45是分页的高度，10是marginTop、marginBottom
		var labelBeforSearch = selectedNode ? selectedNode.value : '请选择左侧数据';
		labelBeforSearch += ': ';
		// 树表中的分页组件，当dialog的宽度低于1000时，设置为大小small， 否则走参照通用大小。
		// 不以分辨率来判断，是因为有在高分辨率下，开发者自定义dialog宽度过小的可能。
		var paginationSize = this.props.dialogWidth < 950 ? 'small' : null;
		var pagination = false;
		//异步树属性
		var lazyProps = {};
		if (this.props.lazytree) {
			lazyProps.loadData = this._loadData;
		}
		if (this.props.pageSize !== -1) {
			pagination = getPagination(totalsize, that.state.current, that.onChangePage, that.onChangePageSize, paginationSize);
		}
		var dataView = React.createElement(
			'div',
			null,
			this.props.customheader,
			React.createElement(
				YYRow,
				{ style: { marginBottom: 10 } },
				React.createElement(
					YYCol,
					{ span: 6, style: { paddingRight: 15, marginTop: -5 } },
					React.createElement(
						YYScrollbar,
						{ style: { height: this.props.contentH - 10 } },
						React.createElement(_YYTreeInRefer2.default, _extends({
							searchFilter: true,
							treeData: treeData,
							loading: this.props.loadingReferTree,
							onTreeChanged: that.onTreeChanged,
							tagNodes: that.props.tagNodes,
							treerelyselect: that.props.treerelyselect,
							storeKeyOfTree: storeKeyOfTree,
							selectedKeys: that.state.selectedNodeKey,
							onSelectNode: that.onSelectNode,
							lazytree: this.props.lazytree,
							emptyView: React.createElement(
								'div',
								{ className: 'yyui-refer-nodata' },
								'\u6682\u65E0\u6570\u636E'
							)
						}, lazyProps))
					)
				),
				React.createElement(
					YYCol,
					{ span: 18, style: { paddingLeft: 20, borderLeftColor: '#e1e1e1', borderLeftWidth: 1, borderLeftStyle: 'solid' } },
					React.createElement(
						'div',
						{ className: 'yyui-refer-treegrid-search' },
						labelBeforSearch,
						React.createElement(YYInputButton, { ref: 'searchGrid', placeholder: that.props.searchPlaceholder, buttonIcon: 'search', onPressEnter: this.onSearchGrid, buttonClick: this.onSearchGrid })
					),
					React.createElement(YYTable, {
						ref: function ref(refComponent) {
							_this3.tableData = refComponent;
						},
						style: { marginTop: 5, width: '100%' },
						columns: columns,
						rowKey: that.props.idField,
						cookieKey: tableKey,
						dataSource: data,
						bordered: true,
						loading: this.props.loadingReferGrid,
						pagination: pagination,
						scroll: { x: totalWidth, y: gridTableH - _ReferConstants.REFER_TABLE_PAGINATION_HEIGHT },
						rowActiveKey: that.state.selectedRowKey,
						onRowDoubleClick: multiselect ? noop : this.onRowDoubleClick,
						onRowClick: multiselect ? noop : this.onRowClick,
						rowSelection: rowSelection,
						rowClickCheckable: this.props.rowClickCheckable })
				)
			)
		);
		var dataTab = React.createElement(
			TabPane,
			{ tab: refname, key: '2', className: 'yyui-refer-tab' },
			dataView
		);
		if (oftenTab === null) {
			return dataView;
		} else {
			return React.createElement(
				Tabs,
				{ defaultActiveKey: '2', onChange: this.changeTags },
				oftenTab,
				dataTab
			);
		}
	},
	render: function render() {
		var that = this;
		var props = {};
		props.refname = this.props.refname;
		props.multiselect = this.props.multiselect;
		props.idField = this.props.idField;
		props.nameField = this.props.nameField;
		props.codeField = this.props.codeField;
		props.onOk = this.btn_ok;
		props.onCancel = this.btn_cancel;
		props.onClean = this.onClean;
		props.onTagClose = this.onTagClose;
		props.selectedDatas = this.state.selectedRows;
		props.refinfokey = this.props.refinfokey;
		props.customButtons = this.props.customButtons;
		return React.createElement(
			'div',
			{ className: this.className() },
			this._renderTabs(),
			React.createElement(_YYReferOperateDlg2.default, { show: that.state.showOperateDialog, operateurl: that.state.operateurl, onCancel: that.onCloseOperate }),
			React.createElement(YYFootInRefer, props)
		);
	}
});

module.exports = (0, _reactRedux.connect)(mapStateToProps, null, null, { withRef: true })(YYReferTreeGrid);
