// yylib-react/dist/yylib2/refer/YYTreeInRefer.js
'use strict';

/**
** 参照中用的树，基于YYTree封装：由YYReferTree、YYReferTreeGrid调用
**/
var React = require('react');
var YYClass = require('../base/YYClass');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
var YYTree = require('../tree/YYTree');
var YYTreeNode = require('../tree/YYTreeNode');
var YYSelect = require('../formcontrols/select/YYSelect');
var ReactUtils = require('../../utils/ReactUtils');
var YYOption = YYSelect.Option;
var store = window.localStorage;

//定义META
var uiMeta = {
    name: 'refer-innertree',
    fields: [
    //field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('treeData', null, PropTypes.array, true, '树数据'), field('checkable', null, PropTypes.bool, true, '是否有复选框'), field('checkedKeys', [], PropTypes.array, true, '默认勾选的项'), field('selectedKeys', [], PropTypes.array, true, '默认选择的项'), field('onCheckNodes', null, PropTypes.func, true, '勾选复选框回调'), field('onSelectNode', null, PropTypes.func, true, '点击树节点回调'), field('treerelyselect', null, PropTypes.array, false, '多树对象数组'), field('tagNodes', null, PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]), false, '默认高亮绿色显示的节点'), field('placeholder', '请输入搜索关键词', PropTypes.string, false, '搜索框的提示信息'), field('lazytree', false, PropTypes.bool, false, '是否启用异步树功能')]
};
var YYTreeInRefer = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        return {
            // treeData: [],
            treeData: this.props.treeData,
            multiple: false,
            checkedKeys: this.props.checkedKeys,
            selectedKeys: this._formatSelectedKey(this.props.selectedKeys),
            loading: this.props.loading,
            expandedKeys: [], // 受控原因：控制默认的初始expandedKeys与用户点击产生的expandedKeys
            autoExpandParent: true
        };
    },

    // 规范selectedKeys
    _formatSelectedKey: function _formatSelectedKey(selectedKeys) {
        if (selectedKeys) {
            if (typeof selectedKeys === "string") {
                //如果是单值时，转化为数组
                return selectedKeys.split(",");
            } else {
                return selectedKeys;
            }
        } else {
            // 如果为空，返回空数组，防止tree内部报null.length错误
            return [];
        }
    },
    componentDidMount: function componentDidMount() {
        // const {treeData} = this.props;
        // setTimeout(() => {
        //     this.setState({
        //       treeData: treeData
        //     });
        // }, 100);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        var _this = this;

        var treeData = nextProps.treeData;

        setTimeout(function () {
            _this.setState({
                treeData: treeData,
                loading: nextProps.loading,
                checkedKeys: nextProps.checkedKeys,
                selectedKeys: _this._formatSelectedKey(nextProps.selectedKeys)
            });
        }, 100);
    },

    //树展开节点事件
    onExpand: function onExpand(expandedKeys) {
        // if not set autoExpandParent to false, if children expanded, parent can not collapse.
        // or, you can remove all expanded children keys.
        this.setState({
            expandedKeys: expandedKeys,
            autoExpandParent: false
        });
    },
    getBoolVal: function getBoolVal(srcVal) {
        if (srcVal === true || srcVal === 'true') {
            return true;
        }
        return false;
    },

    //切换树出发
    onChangeSelect: function onChangeSelect(value) {
        this.props.onTreeChanged(value);
    },

    //树表参照中,切换树情景
    _getSearchSelect: function _getSearchSelect() {
        var that = this;
        var treerelyselect = this.props.treerelyselect;
        if (treerelyselect && treerelyselect.length > 0) {
            var options = [];
            for (var i = 0; i < treerelyselect.length; i++) {
                var curData = treerelyselect[i];
                options.push(React.createElement(
                    YYOption,
                    { value: curData['treerelyurl'], title: curData['relyfieldname'] },
                    curData['relyfieldname']
                ));
            }
            var storeKey = this.props.storeKeyOfTree;
            var firstTree = store[storeKey] ? JSON.parse(store[storeKey]) : treerelyselect[0]['treerelyurl'];
            return React.createElement(
                YYSelect,
                { style: { minWidth: 100 }, defaultValue: firstTree, onChange: that.onChangeSelect },
                options
            );
        } else {
            return null;
        }
    },

    // 合并数组
    _doConcat: function _doConcat(originArray, newArray) {
        if (newArray && newArray instanceof Array && newArray.length > 0) {
            originArray = originArray.concat(newArray);
        }
        return originArray;
    },

    //默认展开根级
    _getDefaultExpandedKeys: function _getDefaultExpandedKeys(treeNodes) {
        //如果是异步树不自动展开节点
        if (this.state.expandedKeys.length > 0) {
            return this.state.expandedKeys;
        }
        var defaultExpandedKeys = [];
        for (var i = 0; i < treeNodes.length; i++) {
            //只有一级跟节点或者 不是异步树时设置
            if (treeNodes[i].props.children && treeNodes[i].props.children.length && (this.props.lazytree == false || treeNodes.length == 1)) {
                treeNodes[i].key ? defaultExpandedKeys.push(treeNodes[i].key) : null;
            }
        }
        // 合并checkedKeys、selectedKeys,以解决默认选择的树节点没有被展开的bug
        var checkedKeys = this.state.checkedKeys;
        //只有一级跟节点或者 不是异步树时设置
        var selectedKeys = this.props.lazytree == false || treeNodes.length == 1 ? this.state.selectedKeys : null;
        defaultExpandedKeys = this._doConcat(defaultExpandedKeys, checkedKeys);
        //如果不是异步树，才将 selectedKeys 拼进默认展开的 defaultExpandedKeys 中
        if (this.props.lazytree === false) defaultExpandedKeys = this._doConcat(defaultExpandedKeys, selectedKeys);
        return defaultExpandedKeys;
    },
    _ifTagNode: function _ifTagNode(item) {
        var tagNodes = this.props.tagNodes;
        return tagNodes && tagNodes.indexOf(item.key) > -1;
    },
    render: function render() {
        var _this2 = this;

        var that = this;
        var loop = function loop(data) {
            return data.map(function (item) {
                var isdisable = _this2.getBoolVal(item.disable);
                // var selectable = this.getBoolVal(!item.unSelectable);
                var selectable = item.selectable === false || item.selectable === 'false' ? false : true;
                var isLeaf = _this2.getBoolVal(item.isLeaf);
                var filterCls = _this2._ifTagNode(item) ? "yyui-refer-tagnode" : '';
                if (item.children) {
                    return React.createElement(
                        YYTreeNode,
                        { className: filterCls, code: item.code, extdata: item.extdata, title: item.name, key: item.key, disabled: isdisable, selectable: selectable, isLeaf: isLeaf },
                        loop(item.children)
                    );
                }
                return React.createElement(YYTreeNode, { className: filterCls, code: item.code, extdata: item.extdata, title: item.name, key: item.key, disabled: isdisable, selectable: selectable, isLeaf: isLeaf });
            });
        };
        var treeNodes = loop(this.state.treeData);
        var expandedKeys = this._getDefaultExpandedKeys(treeNodes);
        var emptyView = this.props.emptyView ? this.props.emptyView : React.createElement(
            'div',
            { style: { marginTop: '30%', marginLeft: '40%' } },
            ' \u65E0\u6570\u636E '
        );
        var tagNodes = this.props.tagNodes;
        if (tagNodes) {
            expandedKeys = expandedKeys ? expandedKeys.concat(tagNodes) : tagNodes;
        };
        return React.createElement(
            YYTree,
            {
                autoExpandParent: this.state.autoExpandParent,
                placeholder: this.props.placeholder,
                className: this.className(),
                loadData: this.props.loadData,
                checkable: this.props.checkable,
                multiple: this.props.multiple,
                loading: this.state.loading,
                expandedKeys: expandedKeys,
                onSelect: this.props.onSelectNode,
                onCheck: this.props.onCheckNodes,
                onExpand: this.onExpand,
                onNodeDblClick: this.props.onNodeDblClick,
                checkedKeys: that.state.checkedKeys,
                selectedKeys: that.state.selectedKeys,
                searchFilter: that.props.searchFilter,
                checkStrictly: that.props.checkStrictly,
                emptyView: emptyView,
                searchAddonBefore: this._getSearchSelect() },
            treeNodes
        );
    }
});

module.exports = YYTreeInRefer;
