// yylib-react/dist/business/enum/YYEnumSelect.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * Created by xiejunb on 2016/12/19.
 */
var React = require('react');

var _require = require('yylib-ui'),
    YYSelect = _require.YYSelect,
    YYMessage = _require.YYMessage,
    YYClass = _require.YYClass;

var field = YYClass.field;
var PropTypes = YYClass.PropTypes;
var DefDocUtil = require('yylib-utils/DefDocUtil');
var AuthToken = require('yylib-utils/AuthToken');
var _ = require('lodash');
var ADDR = require('../BaseHost');
var uiMeta = {
    name: 'enumselect',
    fields: [
    // field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('serverUrl', ADDR, PropTypes.string, true, '服务url'), field('code', null, PropTypes.string, true, '自定义档案类型code'), field('orgId', null, PropTypes.string, false, '组织id'), field('dataparams', null, PropTypes.string, false, '请求数据时的自定义参数'), field('enableOrg', false, PropTypes.bool, false, '是否自动获取当前组织id'), field('valueKey', "code", PropTypes.string, false, '指定下拉选项的value取值,默认取自定义档案的id,可取值id、code'), field('textKey', "name", PropTypes.string, false, '指定下拉选项的text取值'), field('isRefreshCache', false, PropTypes.bool, false, '是否刷新缓存,默认false不刷新'), field('multiple', false, PropTypes.bool, false, '支持多选'), field('allowClear', false, PropTypes.bool, false, '支持清除, 单选模式有效'), field('tags', false, PropTypes.bool, false, '可以把随意输入的条目作为 tag，输入项不需要与下拉选项匹配'), field('dropdownMatchSelectWidth', true, PropTypes.bool, false, '下拉菜单和选择器同宽'), field('combobox', false, PropTypes.bool, false, '输入框自动提示模式'), field('showSearch', false, PropTypes.bool, false, '在选择框中显示搜索框'), field('onShowFilter', null, PropTypes.func, false, '自定义函数过滤出允许显示到下拉列表的项目,注意return允许显示的下拉项数组'), field('disabled', false, PropTypes.bool, false, '是否禁用'), field('defaultActiveFirstOption', false, PropTypes.bool, false, '是否默认高亮第一个选项')
    // , field('labelInValue', false, PropTypes.bool, false, '是否把每个选项的 label 包装到 value 中，决定 Select 的 value 类型')
    , field('onSelect', null, PropTypes.func, false, '被选中时调用，参数为选中项的 value 值'), field('onDeselect', null, PropTypes.func, false, '取消选中时调用，参数为选中项的 option value 值，仅在 multiple 或 tags 模式下生效'), field('onChange', null, PropTypes.func, false, '选中 option，或 input 的 value 变化（combobox 模式下）时，调用此函数'), field('onSearch', null, PropTypes.func, false, '文本框值变化时回调'), field('value', undefined, PropTypes.oneOf([PropTypes.string, PropTypes.array]), false, '指定当前选中的条目'), field('defaultValue', undefined, PropTypes.oneOf([PropTypes.string, PropTypes.array]), false, '指定当前选中的条目'), field('filterOption', undefined, PropTypes.oneOf([PropTypes.bool, PropTypes.func]), false, '是否根据输入项进行筛选。当其为一个函数时，会接收 inputValue option 两个参数，当 option 符合筛选条件时，应返回 true，反之则返回 false。'), field('size', 'default', PropTypes.oneOf(['large', 'default', 'small']), false, '选择框大小，可选 large small'), field('placeholder', '', PropTypes.string, false, '选择框默认文字'), field('loadingContent', '正在加载数据', PropTypes.string, false, '请求数据加载时下拉列表显示的内容'), field('notFoundContent', '暂无相关数据', PropTypes.string, false, '当下拉列表为空时显示的内容'), field('optionFilterProp', 'value', PropTypes.string, false, '搜索时过滤对应的 option 属性，如设置为 children 表示对内嵌内容进行搜索'), field('optionLabelProp', undefined, PropTypes.string, false, '选择框回填到选择框的 Option 的属性值，默认是 Option 的子元素。比如在子元素需要高亮效果时，此值可以设为 value默认文字'), field('getPopupContainer', function () {
        return document.body;
    }, PropTypes.func, false, '菜单渲染父节点。默认渲染到 body 上，如果你遇到菜单滚动定位问题，试试修改为滚动的区域，并相对其定位'), field('propKey', 'code', PropTypes.oneOf([PropTypes.string, PropTypes.array]), false, '需要加到对象value里的属性'), field('valueType', 'string', PropTypes.string, false, '值类型string, object')]
};
var YYEnumSelect = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        var props = this.props;
        var value = props.value ? props.value : props.defaultValue;
        return {
            options: [],
            value: value,
            dropdownlistContent: props.loadingContent
        };
    },
    componentDidMount: function componentDidMount() {
        var orgId = this._getOrgId(this.props.orgId);
        this._requestData(this.props.code, orgId, this.props.dataparams, this.props.isRefreshCache);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('defaultValue' in nextProps) {
            if (this.props.defaultValue != nextProps.defaultValue) {
                this.setState({ value: nextProps.defaultValue });
            }
        }
        //value优先级高于defaultValue
        if ('value' in nextProps) {
            if (this.props.value != nextProps.value) {
                this.setState({ value: nextProps.value });
            }
        }
        if (nextProps.code !== this.props.code) {
            var orgId = this._getOrgId(nextProps.orgId);
            this.setState({
                dropdownlistContent: nextProps.loadingContent
            });
            this._requestData(nextProps.code, orgId, nextProps.dataparams, nextProps.isRefreshCache);
        }
        if (nextProps.dataparams !== this.props.dataparams) {
            var orgId = this._getOrgId(nextProps.orgId);
            this.setState({
                dropdownlistContent: nextProps.loadingContent
            });
            this._requestData(nextProps.code, orgId, nextProps.dataparams, nextProps.isRefreshCache);
        }
        if ('isRefreshCache' in nextProps && nextProps.isRefreshCache) {
            var orgId = this._getOrgId(nextProps.orgId);
            this.setState({
                dropdownlistContent: nextProps.loadingContent
            });
            this._requestData(nextProps.code, orgId, nextProps.dataparams, nextProps.isRefreshCache);
        }
    },
    _getOrgId: function _getOrgId(currentOrgId) {
        var orgId = null;
        if (this.props.enableOrg) {
            orgId = AuthToken.getOrgaId();
        }
        if (currentOrgId) {
            orgId = currentOrgId;
        }
        return orgId;
    },
    _requestData: function _requestData(code, orgId, dataparams, isRefreshCache) {
        if (!code) return; //不存在则不执行查询
        DefDocUtil.findDefDocListByCode(this.props.serverUrl, code, this._callback, orgId, dataparams, isRefreshCache);
    },
    _callback: function _callback(result) {
        // var valueKey = this.props.valueKey;
        // if(valueKey==='id') {
        //     valueKey = 'defdocId';
        // }
        if (result.success) {
            // var defDocList = result.backData;
            // var options = defDocList.map(function (defDoc) {
            //     var option = {value: defDoc[valueKey], text: defDoc.name}
            //     return option;
            // });
            this.setState({ options: result.backData });
        } else {
            YYMessage.error(result.backMsg);
        }
        this.setState({
            dropdownlistContent: this.props.notFoundContent
        });
    },
    _onChange: function _onChange(keys) {
        if (_.isArray(keys)) {
            keys = _.map(keys, function (item) {
                if (_.isObject(item) && item.defdocId) {
                    item.id = item.defdocId;
                    delete keys['defdocId'];
                }
                return item;
            });
        } else if (_.isObject(keys) && keys.defdocId) {
            keys.id = keys.defdocId;
            delete keys['defdocId'];
        }
        var that = this;
        this.setState({ value: keys }, function () {
            if (_.isFunction(that.props.onChange)) {
                that.props.onChange(keys);
            }
        });
    },
    render: function render() {
        var _props = this.props,
            propKey = _props.propKey,
            valueKey = _props.valueKey;

        var value = this.state.value;
        if (valueKey === 'id') {
            valueKey = 'defdocId';
            if (_.isArray(value)) {
                //多选模式下
                value = _.map(value, function (val) {
                    if (_.isObject(val)) {
                        val.defdocId = val['id'] ? val['id'] : val.defdocId;
                    }
                    return val;
                });
            } else if (_.isObject(value)) {
                //单选模式下
                value.defdocId = value['id'] ? value['id'] : value.defdocId;
            }
        }
        return React.createElement(YYSelect, _extends({}, this.props, { notFoundContent: this.state.dropdownlistContent, propKey: propKey, valueKey: valueKey, items: this.state.options, className: this.className(),
            onChange: this._onChange, value: value }));
    }
});

module.exports = YYEnumSelect;
