// yylib-react/dist/yylib2/formcontrols/datepicker/YYYearCalendar.js
'use strict';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var YearPanel = require('rc-calendar/lib/year/YearPanel'); /**
                                                           *** by zhangzhzhc 2017/04/11
                                                           *** 年 日历组件。 基于rc-calendar中 YearPanel、CalendarMixin、CommonMixin抽取
                                                           **/

var CalendarMixin = require('rc-calendar/lib/mixin/CalendarMixin');
var CommonMixin = require('rc-calendar/lib/mixin/CommonMixin');
require('rc-calendar/assets/index.css'); //必需，勿删
require('./style/index.less');

//因为必须用到mixins，所以不能采用ES6写法
var YYYearCalendar = _react2.default.createClass({
    displayName: 'YYYearCalendar',

    mixins: [CommonMixin, CalendarMixin],
    render: function render() {
        var that = this;
        return _react2.default.createElement(
            'div',
            { className: 'ant-calendar-picker-container ant-calendar-picker-container-placement-topLeft' },
            _react2.default.createElement(
                'div',
                { className: 'ant-calendar ant-calendar-year' },
                _react2.default.createElement(YearPanel, {
                    value: that.state.value,
                    rootPrefixCls: 'ant-calendar',
                    onSelect: that.props.onSelect,
                    locale: that.props.locale
                })
            )
        );
    }
});

module.exports = YYYearCalendar;
