// yylib-react/dist/business/bpmapprove/bpm/constants/ActionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
//提交
var BPM_SUBMIT_SUCCESS = exports.BPM_SUBMIT_SUCCESS = 'BPM_SUBMIT_SUCCESS';
var BPM_SUBMIT_ERROR = exports.BPM_SUBMIT_ERROR = 'BPM_SUBMIT_ERROR';
//审批
var BPM_APPROVE_SUCCESS = exports.BPM_APPROVE_SUCCESS = 'BPM_APPROVE_SUCCESS';
var BPM_APPROVE_ERROR = exports.BPM_APPROVE_ERROR = 'BPM_APPROVE_ERROR';
//审批前查询
var BPM_PREAPPROVEQUERY_SUCCESS = exports.BPM_PREAPPROVEQUERY_SUCCESS = 'BPM_PREAPPROVEQUERY_SUCCESS';
var BPM_PREAPPROVEQUERY_ERROR = exports.BPM_PREAPPROVEQUERY_ERROR = 'BPM_PREAPPROVEQUERY_ERROR';
//收回
var BPM_CALLBACK_SUCCESS = exports.BPM_CALLBACK_SUCCESS = 'BPM_CALLBACK_SUCCESS';
var BPM_CALLBACK_ERROR = exports.BPM_CALLBACK_ERROR = 'BPM_CALLBACK_ERROR';
//驳回前查询
var BPM_PREREJECT_SUCCESS = exports.BPM_PREREJECT_SUCCESS = 'BPM_PREREJECT_SUCCESS';
var BPM_PREREJECT_ERROR = exports.BPM_PREREJECT_ERROR = 'BPM_PREREJECT_ERROR';
//驳回
var BPM_REJECT_SUCCESS = exports.BPM_REJECT_SUCCESS = 'BPM_REJECT_SUCCESS';
var BPM_REJECT_ERROR = exports.BPM_REJECT_ERROR = 'BPM_REJECT_ERROR';
//查看审批流
var BPM_VIEWAPPROVEAL_SUCCESS = exports.BPM_VIEWAPPROVEAL_SUCCESS = 'BPM_VIEWAPPROVEAL_SUCCESS';
var BPM_VIEWAPPROVEAL_ERROR = exports.BPM_VIEWAPPROVEAL_ERROR = 'BPM_VIEWAPPROVEAL_ERROR';
//待办
var BPM_TODOTASK_SUCCESS = exports.BPM_TODOTASK_SUCCESS = 'BPM_TODOTASK_SUCCESS';
var BPM_TODOTASK_ERROR = exports.BPM_TODOTASK_ERROR = 'BPM_TODOTASK_ERROR';
//已办
var BPM_DONETASK_SUCCESS = exports.BPM_DONETASK_SUCCESS = 'BPM_DONETASK_SUCCESS';
var BPM_DONETASK_ERROR = exports.BPM_DONETASK_ERROR = 'BPM_DONETASK_ERROR';
//改派
var BPM_DELEGATETASK_SUCCESS = exports.BPM_DELEGATETASK_SUCCESS = 'BPM_DELEGATETASK_SUCCESS';
var BPM_DELEGATETASK_ERROR = exports.BPM_DELEGATETASK_ERROR = 'BPM_DELEGATETASK_ERROR';
//弃审
var BPM_UNAPPROVE_SUCCESS = exports.BPM_UNAPPROVE_SUCCESS = 'BPM_UNAPPROVE_SUCCESS';
var BPM_UNAPPROVE_ERROR = exports.BPM_UNAPPROVE_ERROR = 'BPM_UNAPPROVE_ERROR';
//不同意且退回
var BPM_DISAPPROVE_SUCCESS = exports.BPM_DISAPPROVE_SUCCESS = 'BPM_DISAPPROVE_SUCCESS';
var BPM_DISAPPROVE_ERROR = exports.BPM_DISAPPROVE_ERROR = 'BPM_DISAPPROVE_ERROR';
//查询我提交的
var BPM_MYSUBMIT_SUCCESS = exports.BPM_MYSUBMIT_SUCCESS = 'BPM_MYSUBMIT_SUCCESS';
var BPM_MYSUBMIT_ERROR = exports.BPM_MYSUBMIT_ERROR = 'BPM_MYSUBMIT_ERROR';
//获取当前登录信息
var BPM_GETUSER_SUCCESS = exports.BPM_GETUSER_SUCCESS = 'BPM_GETUSER_SUCCESS';
var BPM_GETUSER_ERROR = exports.BPM_GETUSER_ERROR = 'BPM_GETUSER_ERROR';
//获取审批指派用户
var BPM_APPROVEASSIGNUSER_SUCCESS = exports.BPM_APPROVEASSIGNUSER_SUCCESS = 'BPM_APPROVEASSIGNUSER_SUCCESS';
var BPM_APPROVEASSIGNUSER_ERROR = exports.BPM_APPROVEASSIGNUSER_ERROR = 'BPM_APPROVEASSIGNUSER_ERROR';
//获取提交指派用户
var BPM_STARTASSIGNUSER_SUCCESS = exports.BPM_STARTASSIGNUSER_SUCCESS = 'BPM_STARTASSIGNUSER_SUCCESS';
var BPM_STARTASSIGNUSER_ERROR = exports.BPM_STARTASSIGNUSER_ERROR = 'BPM_STARTASSIGNUSER_ERROR';
//获取配置的流程信息
var BPM_GETCONFIGLIST_SUCCESS = exports.BPM_GETCONFIGLIST_SUCCESS = 'BPM_GETCONFIGLIST_SUCCESS';
var BPM_GETCONFIGLIST_ERROR = exports.BPM_GETCONFIGLIST_ERROR = 'BPM_GETCONFIGLIST_ERROR';
