// yylib-react/dist/business/attach/ImageUpload.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * 图片上传组件
 * 本组件已过期，请使用YYImageWall组件
 */
var React = require('react');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass,
    YYPage = _require.YYPage,
    YYUpload = _require.YYUpload,
    YYMessage = _require.YYMessage,
    YYButton = _require.YYButton,
    YYIcon = _require.YYIcon,
    YYTooltip = _require.YYTooltip,
    YYNotice = _require.YYNotice;

var _require2 = require("./attachaction"),
    loadAttachList = _require2.loadAttachList,
    delAttach = _require2.delAttach,
    downloadUrl = _require2.downloadUrl,
    listUrl = _require2.listUrl,
    uploadUrl = _require2.uploadUrl,
    delUrl = _require2.delUrl;

var _require3 = require('lodash'),
    isFunction = _require3.isFunction;

var ADDR = require('../BaseHost');
//var {Upload} = require("antd");
//下载url
//const downloadUrl = "/icop-file/file/download?id=";
//获取附件列表的url
//const listUrl = "/icop-file/file/list";
//上传url
//const uploadUrl = "/icop-file/file/upload2";
//const delUrl = "/icop-file/file/del";

var ImageUpload = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        //serverUrl为上传接口服务url前缀，fileUrl为文件访问url前缀
        return { serverUrl: ADDR, fileUrl: ADDR, value: '', height: 100, width: 100, sourceId: '', sourceType: '', billType: '', title: '附件上传', disabled: false };
    },
    getInitialState: function getInitialState() {
        var value = null;
        if (this.props.__meta != undefined && this.props.__meta.initialValue != undefined) {
            value = this.props.__meta.initialValue;
        } else if (this.props.value != undefined) {
            value = this.props.value;
        } else {
            value = this.props.value;
        }
        return {
            value: value, fileList: [], isload: false,
            fileid: '', sourceId: '', filePath: '', //sourceId用于控制加载次数
            disabled: false
        };
    },

    componentWillMount: function componentWillMount() {
        //可编辑表格会先加载子组件upload，调用onchange会传递给父组件，此时YYEditTable还没构建才完成所以会出bug
        /*if(isFunction(this.props.onChange)){
            this.props.onChange(null);
        }*/
    },
    _loadImg: function _loadImg(serverUrl, billType, sourceId, sourceType) {
        var url = serverUrl + listUrl;
        var that = this;
        loadAttachList(url, { id: sourceId, billType: billType, type: sourceType }, function (data) {
            if (data) {
                var fileid = data[0].gid;
                var filepath = data[0].filePath;
                that.setState({ fileid: fileid, filePath: filepath });
            } else {
                that.setState({ value: null, fileList: [], fileid: '', filePath: '' });
            }
        });
    },
    componentDidMount: function componentDidMount() {
        //由于只有在sourceId为固定值时才会调用该区域，而现实情况不会有这种情况，所以暂时注释掉
        /* if(this.state.isload){
             return;
         }*/
        var sourceId = this.props.sourceId;
        var billType = this.props.billType;
        var disabled = this.props.disabled;
        this.setState({ disabled: disabled });
        if (sourceId && billType && this.state.isload == false) {
            this.setState({ isload: true, sourceId: sourceId });

            var serverUrl = this.props.serverUrl;
            var sourceType = this.props.sourceType;
            this._loadImg(serverUrl, billType, sourceId, sourceType);
        }
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        /*if(this.state.isload){
            return;
        }*/
        var sourceId = nextProps.sourceId;
        var billType = nextProps.billType;
        var lastSourceId = this.state.sourceId;
        var disabled = nextProps.disabled;
        this.setState({ disabled: disabled });
        if (!sourceId) {
            sourceId = "";
        }
        if (!lastSourceId) {
            lastSourceId = "";
        }
        if (sourceId && billType) {
            this.setState({ isload: true });
            if (this.state.sourceId != sourceId) {
                //sourceId变化时才加载图片
                var serverUrl = nextProps.serverUrl;
                var sourceType = nextProps.sourceType;
                this._loadImg(serverUrl, billType, sourceId, sourceType);
            }
        } else {
            if (sourceId != lastSourceId) {
                //如果sourceId改变了，则清空数据
                this.setState({ value: null, fileList: [], fileid: '', filePath: '' });
            }
        }
        this.setState({ sourceId: sourceId });
    },
    handleChange: function handleChange(info) {
        var that = this;
        if (info.file.status === 'done') {
            that.setState({ fileList: [] });
            var back = info.file.response;
            if (back.success) {
                var attachArr = back.backData;
                var newFileid = attachArr[0].gid;
                var newFilePath = attachArr[0].filePath;
                if (that.props.sourceId) {
                    //修改时，先上传后删除
                    var sourceId = that.props.sourceId;
                    var billType = that.props.billType;
                    var sourceType = that.props.sourceType;
                    //删除旧的文件
                    var ids = that.state.fileid;
                    //图片更新为新的链接
                    that.setState({ fileid: newFileid, filePath: newFilePath });

                    delAttach(this.props.serverUrl + delUrl, { id: sourceId, billType: billType, sourceType: sourceType, attachIds: ids }, function (data) {
                        if (data.success) {
                            if (console) {
                                console.log("删除成功");
                            }
                        }
                    }, function (err) {
                        if (console) {
                            console.log("删除失败" + err);
                        }
                    });
                } else {
                    //新增时，直接更新组件的value
                    that.setState({ fileid: newFileid, filePath: newFilePath });
                    if (isFunction(that.props.onChange)) {
                        that.props.onChange(newFileid);
                    }
                }
            }
        }
    },
    beforeUpload: function beforeUpload(file) {
        if (file.size == 0) {
            showErrMsg('上传失败', '请不要上传空文件');
            return false;
        }
    },
    staticUrl: function staticUrl() {
        if (this.props.fileUrl) {
            return this.props.fileUrl;
        }
        return this.props.serverUrl;
    },
    image: function image() {
        var width = parseInt(this.props.width);
        var height = parseInt(this.props.height);
        //if(this.state.fileid){
        //console.log("_"+width+"x"+height);
        if (this.state.filePath) {
            var staticUrl = this.staticUrl();
            var url = staticUrl + "/" + this.state.filePath;
            var orginUrl = url;
            //缩略图尺寸
            var dimension = "_" + width + "x" + height;
            var idx = url.lastIndexOf(".");
            if (idx > -1) {
                var ext = url.substr(idx);
                url = url.substr(0, idx) + dimension + ext;
                console.log(url);
            }
            return React.createElement(
                'div',
                { style: { 'border': '1px dashed #d9d9d9', 'border-radius': '6px', 'height': height + 5, 'width': width + 5 } },
                React.createElement(
                    'a',
                    { target: '_blank', href: orginUrl },
                    React.createElement('img', { width: width, height: height, src: url })
                )
            );
        } else {
            //没有上传照片时
            return React.createElement(
                'div',
                { className: 'ant-upload ant-upload-select ant-upload-select-picture-card', width: width, height: height },
                React.createElement(YYIcon, { type: 'plus' }),
                React.createElement(
                    'div',
                    { className: 'ant-upload-text' },
                    '\u8BF7\u5148\u4E0A\u4F20\u7167\u7247'
                )
            );
        }
    },
    onUpload: function onUpload(e) {
        e.preventDefault();
    },
    render: function render() {
        var sourceId = this.props.sourceId;
        var billType = this.props.billType;
        var sourceType = this.props.sourceType;
        var that = this;
        var props = {
            action: this.props.serverUrl + uploadUrl,
            data: { sourceId: sourceId, billType: billType, sourceType: sourceType },
            onChange: this.handleChange,
            showUploadList: false,
            multiple: false,
            beforeUpload: this.beforeUpload
        };
        var title = this.props.title ? this.props.title : '点击上传';
        return React.createElement(
            'div',
            null,
            this.image(),
            this.state.disabled ? React.createElement('div', null) : React.createElement(
                'div',
                null,
                React.createElement(
                    YYUpload,
                    _extends({}, props, { accept: 'image/gif,image/jpeg,image/jpg,image/png' }),
                    React.createElement(
                        YYTooltip,
                        { placement: 'top', title: '\u8BF7\u9009\u62E9\u56FE\u7247\u4E0A\u4F20' },
                        React.createElement(
                            YYButton,
                            { type: 'ghost', onClick: that.onUpload },
                            React.createElement(YYIcon, { type: 'upload' }),
                            title
                        )
                    )
                )
            )
        );
    }
});

module.exports = ImageUpload;
