// yylib-react/dist/business/templates/resource/MDModuleData.js
'use strict';

var React = require('react');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass;

var AceEditor = require('react-ace').default;
var ReduxUtils = require('yylib-utils/ReduxUtils');
var TreeDataUtils = require('../../../../apps/quickdev/utils/TreeDataUtils');

var _require2 = require('../../../../apps/quickdev/data/UiDef'),
    UiTypeDef = _require2.UiTypeDef;

var mapStateToProps = function mapStateToProps(state) {
    return {
        moduleData: TreeDataUtils.findNodeByUiType(state.ide.resourceData.children, UiTypeDef.md)
    };
};

var MDModuleData = YYClass.create({
    onEditCodeChange: function onEditCodeChange() {},
    render: function render() {
        return React.createElement(AceEditor, {
            mode: 'javascript',
            theme: 'monokai',
            fontSize: 14,
            onChange: this.onEditCodeChange,
            name: 'moduledata',
            width: '100%',
            height: '100%',
            showPrintMargin: false,
            value: JSON.stringify(this.props.moduleData, null, '\t'),
            editorProps: { $blockScrolling: true }
        });
    }
});

module.exports = ReduxUtils.connect(MDModuleData, mapStateToProps);
