// yylib-cscec/dist/template/export/ExportExcelModal.js
'use strict';

/**
 * Created by xg on 2017/7/6.
 * 导出Excel时确认导出数据范围对话框
 */
var React = require('react');

var _require = require('yylib-ui'),
    YYClass = _require.YYClass,
    YYModal = _require.YYModal,
    YYRadioGroup = _require.YYRadioGroup,
    YYRadio = _require.YYRadio;

var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
var _ = require('lodash');
var PubConstants = require('../resource/PubConstants');

var uiMeta = {
    name: 'exportExcel',
    fields: [
    //field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('visible', false, PropTypes.bool, false, '是否显示对话框'), field('onExport', undefined, PropTypes.func, false, '导出按钮点击事件回调函数'), field('onCancel', undefined, PropTypes.func, false, '取消按钮点击事件回调函数')]
};
var ExportExcelModal = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        return {
            visible: this.props.visible,
            selectValue: '' + PubConstants.EXPORT.SCOPE.PAGE
        };
    },

    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ("visible" in nextProps) {
            this.setState({ visible: nextProps.visible });
        }
    },


    /**
     * 导出按钮点击事件处理
     */
    onExport: function onExport() {
        if (this.props.onExport && _.isFunction(this.props.onExport)) {
            this.props.onExport(this.state.selectValue);
        } else {
            console.warn("onExport回调事件未定义");
        }
    },

    /**
     * 取消按钮点击事件处理
     */
    onCancel: function onCancel() {
        if (this.props.onCancel && _.isFunction(this.props.onCancel)) {
            this.props.onCancel();
        }
    },

    /**
     * 单选按钮切换事件
     * @param e
     */
    onSelectChange: function onSelectChange(e) {
        this.setState({ selectValue: e.target.value });
    },

    render: function render() {
        var radioStyle = {
            display: 'block',
            height: '30px',
            lineHeight: '30px',
            marginLeft: '30px',
            marginTop: '5px',
            marginBottom: '5px',
            fontSize: '14px'
        };
        return React.createElement(
            YYModal,
            { width: '400', title: '\u8BF7\u9009\u62E9\u6570\u636E\u8303\u56F4', okText: '\u5BFC\u51FA', visible: this.state.visible, onOk: this.onExport,
                onCancel: this.onCancel, maskClosable: false },
            React.createElement(
                YYRadioGroup,
                { onChange: this.onSelectChange, value: this.state.selectValue },
                React.createElement(
                    YYRadio,
                    { key: '' + PubConstants.EXPORT.SCOPE.LINE,
                        value: '' + PubConstants.EXPORT.SCOPE.LINE, style: radioStyle },
                    '\u9009\u4E2D\u884C\u6570\u636E'
                ),
                React.createElement(
                    YYRadio,
                    { key: '' + PubConstants.EXPORT.SCOPE.PAGE,
                        value: '' + PubConstants.EXPORT.SCOPE.PAGE, style: radioStyle },
                    '\u5F53\u524D\u9875\u9762\u6570\u636E'
                ),
                React.createElement(
                    YYRadio,
                    { key: '' + PubConstants.EXPORT.SCOPE.ALL,
                        value: '' + PubConstants.EXPORT.SCOPE.ALL, style: radioStyle },
                    '\u6240\u6709\u6570\u636E'
                )
            )
        );
    }
});
module.exports = ExportExcelModal;
