// yylib-react/dist/business/bpmapprove/bpm/url/BPMURL.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
//项目名称
var MOD_PUBBASE_BPMCENTER = '/icop-bpmcenter-web/';
var BPMURL = {
  BPM_SUBMIT: "" + MOD_PUBBASE_BPMCENTER + 'bpm/doSubmit', //提交
  BPM_APPROVE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/doApprove', //审批
  BPM_BATCHAPPROVE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/doBatchApprove', //审批
  BPM_PREAPPROVEQUERY: "" + MOD_PUBBASE_BPMCENTER + 'bpm/beforeApprove', //审批前查询
  BPM_CALLBACK: "" + MOD_PUBBASE_BPMCENTER + 'bpm/doCallBack', //收回
  BPM_PREREJECT: "" + MOD_PUBBASE_BPMCENTER + 'bpm/beforeReject', //驳回前查询
  BPM_REJECT: "" + MOD_PUBBASE_BPMCENTER + 'bpm/doReject', //驳回
  BPM_VIEWAPPROVEAL: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryHistoricProcessInfo', //查看审批流
  BPM_New_VIEWAPPROVEAL: "" + MOD_PUBBASE_BPMCENTER + 'bpm/querySingleHistoricProcessInfo', //新查看审批流
  BPM_VIEWAPPROVEAL_VIEWURL: "" + MOD_PUBBASE_BPMCENTER + 'pages/diagram-viewer',
  BPM_TODOTASK: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryTodoTask', //待办
  BPM_DONETASK: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryDoneTask', //已办
  BPM_DELEGATE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/delegateTaskCompletely', //改派
  BPM_UNAPPROVE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/unapprove', //弃审
  BPM_DISAPPROVE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/disApprove', //不同意且退回
  BPM_MYSUBMIT: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryMySubmit', //获取我提交的
  BPM_GETUSER: "" + MOD_PUBBASE_BPMCENTER + 'bpm/getUser', //获取用户
  BPM_APPROVEASSIGNUSER: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryAssignUser', //获取审批指派用户
  BPM_STARTASSIGNUSER: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryStartAssignUser', //获取审批指派用户
  BPM_GETASSIGNBYPROCKEY: "" + MOD_PUBBASE_BPMCENTER + 'bpm/queryAssignUserByKey',
  BPM_GETCONFIGLIST: "" + MOD_PUBBASE_BPMCENTER + 'processConfig/getConfigList',
  BPM_UPDATEBILLSTATE: "" + MOD_PUBBASE_BPMCENTER + 'bpm/updateBillState'
};
exports.default = BPMURL;
