// yylib-react/dist/business/searchfilter/YYFilterToggle.js
'use strict';

/**
 * Created by Dio on 2016/5/11.
 * 收起|展开
 */
var React = require('react');

var _require = require('lodash'),
    isFunction = _require.isFunction;

var _require2 = require('yylib-ui'),
    YYClass = _require2.YYClass,
    YYCol = _require2.YYCol,
    YYIcon = _require2.YYIcon;

var YYFilterToggle = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        return {
            expand: true //是否展开
            , expandText: '收起筛选' //展开时显示的文本
            , collapseText: '展开筛选' //折叠时显示的文本
            , expandIcon: "up" //展开时的图标样式名
            , collapseIcon: "down" //折叠时的图标样式名
            , onExpandOrCollapse: null //当展开|折叠时的回调，传递参数：isExpand(true-展开，false-折叠)
        };
    },
    getInitialState: function getInitialState() {
        return {
            expand: this.props.expand == true || this.props.expand == 'true'
        };
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('expand' in nextProps) {
            if (this.state.expand != nextProps.expand) {
                this.setState({ expand: nextProps.expand });
            }
        }
    },
    _onChange: function _onChange() {
        var isExpand = !this.state.expand;
        this.setState({ expand: isExpand });
        if (isFunction(this.props.onExpandOrCollapse)) {
            this.props.onExpandOrCollapse(isExpand);
        }
    },
    render: function render() {
        var cls = "yyui-filter-toggle " + (this.state.expand ? "" : "collapse");
        var icon = this.state.expand ? React.createElement(YYIcon, { type: this.props.expandIcon }) : React.createElement(YYIcon, { type: this.props.collapseIcon });
        return React.createElement(
            'div',
            { className: cls, onClick: this._onChange },
            React.createElement(
                YYCol,
                { span: '22' },
                React.createElement(
                    'span',
                    null,
                    this.state.expand ? this.props.expandText : this.props.collapseText
                )
            ),
            React.createElement(
                YYCol,
                { span: '2', style: { marginLeft: this.props.collapseIcon ? '-10px' : '0px' }, className: 'yyui-more' },
                icon
            )
        );
    }
});
//规范属性类型
var propTypes = {
    expand: React.PropTypes.bool,
    onExpandOrCollapse: React.PropTypes.function
};
YYFilterToggle.PropTypes = propTypes;

module.exports = YYFilterToggle;
